---
title: "Algebra and Statistical Methods for Big Data witn HDF5 files"
author: 
- name: Dolors Pelegri
  affiliation: 
  - &uab Universitat Autonoma de Barcelona - Universitat Autonoma de Barcelona (UAB)
  - &isglobal ISGlobal, Centre for Research in Environmental Epidemiology (CREAL)
  - &brge Bioinformatics Research Group in Epidemiolgy (BRGE)
- name: Juan R. Gonzalez
  affiliation: 
  - *uab
  - *isglobal
  - *brge
  email: juanr.gonzalez@isglobal.org
date: "`r Sys.Date()`"
package: "`r pkg_ver('BigDataStatMeth')`"
output:
  BiocStyle::html_document: default
  BiocStyle::pdf_document:
    toc_float: true
abstract: |
  Description of functions to perform matrix operations using HDF5 data files.
vignette: |
  %\VignetteIndexEntry{Algebra and Statistical Methods for Big Data witn HDF5 files}
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Integrating TCGA Data}
  %\VignetteEngine{knitr::rmarkdown}
bibliography: references.bib
---

```{r setup, include = FALSE}
library(knitr)
library(BiocStyle)

knitr::opts_chunk$set(collapse = TRUE, comment = "", cache = TRUE, message = FALSE, width = 180)
```

```{r, cleanup, echo=FALSE, include=FALSE}
if( isTRUE(file.exists('delayed.hdf5'))) {
    file.remove('delayed.hdf5')
}
if( isTRUE(file.exists('robjects.hdf5'))){
    file.remove('robjects.hdf5')
}
if( isTRUE(file.exists('colesterol_file.hdf5'))){
    file.remove('colesterol_file.hdf5')
}
if( isTRUE(file.exists('tmp_blockmult.hdf5'))){
    file.remove('tmp_blockmult.hdf5')
}
```

<!--
```{r r_setup, include = FALSE}
options(width = 180)
```
-->
# Prerequisites

The package requires other `Bioconductor` packages to be installed. These include: `beachmat`, `HDF5Array`,
`DelayedArray`, `rhdf5`. Other required R packages are: `Matrix`,
`RcppEigen` and `RSpectra`. 

# Overview

This package implements seveal matrix operations using `DelayMatrix` objects and HDF5 data files as well as some basic algebra operations that can be used to carry out some statistical analyses using standard methodologies such as principal component analyses or least squares estimation. The package also contains specific statatistical method mainly used in `omic` data analysis such as lasso regression. 

# Getting started

First, let us start by loading the required packages to describe the main capabilities of the package

```{r, load}
library(BigDataStatMeth)
library(rhdf5)
library(DT)
```

These other packages are required to reproduce this vignette

```{r, load2}
library(microbenchmark)
```

# Previous knowledge

## HDF5 data files

Hierarchical Data Format (HDF) is a set of file formats  designed to store and organize large amounts of data. It is supported by The HDF Group, a non-profit corporation whose mission is to ensure continued development of HDF5 technologies and the continued accessibility of data stored in HDF.

HDF5 is a technology suite that makes possible the management of extremely large and complex data collections, can accomodate virtually every kind of data in a single file, sequences, images, SNP matrices, and every other type of data and metadata associated with an experiment.

There is no limit on the number or size of data objects in the collection, giving great flexibility for omic data. Is high-performance I/O with a rich set of integrated performance features that allow for access time and storage space optimizations

HDF5 file structure include only *two major types of object*:

- `Datasets`, which are multidimensional arrays of a homogeneous type, in datasets we can store omics data  associated to genomics, transcriptomics, epigonomics, proteomics, metabolomics experiments

- `Groups`, which are container structures which can hold datasets and other groups

This results in a truly hierarchical, filesystem-like data format

```{r hdf5Img, out.width = '100%', fig.align = 'center', fig.cap = "HDF5 hierarchical structure", echo=FALSE}
knitr::include_graphics("imgs/hdf5_squema.jpg")
```


## Basics in rhdf5

**Create hdf5 file**

The `Create_HDF5_matrix_file` function implemented in this package creates an hdf5 file with a group and a dataset in one step. This function allows to create datasets from `DelayedMAtrix` objects using `beachmat` library as well as standard R matrices from matrix object.

```{r hdf5Create}
library(rhdf5)

set.seed(5234)
n <- 500
m <- 600
A <- matrix(rnorm(n*m,mean=0,sd=1), n,m)

# We also can create a dataset from R matrix object
Create_HDF5_matrix_file("robjects.hdf5", A, "INPUT", "A")
```

**Add datasets in hdf5 file**

The `Create_HDF5_matrix` allows to create a dataset in existing file. We can create the dataset in any group, if group doesn't exists in file, the group is created before append the dataset. The object could be an R object or a DelayedArray object.

```{r hdf5AddDataset}

set.seed(5234)
n <- 100
m <- 10000
A <- matrix(rnorm(n*m,mean=0,sd=1), n,m)

set.seed(5234)
n <- 50
m <- 12000
B <- matrix(rnorm(n*m,mean=3,sd=0.5), n,m)

# Now, we create anothre data file with matrix A naemed delayed.hdf5
Create_HDF5_matrix_file("delayed.hdf5", A, "INPUT", "A");

# And now, we create dataframe B with B matrix in delayed.hdf5 file at INPUT group
Create_HDF5_matrix(B, "delayed.hdf5", "INPUT", "B");

```

**Create transposed matrix in hdf5 file**

The function `Create_HDF5_matrix` allows to create a dataset with a transposed matrix with the parameter `transpose = TRUE``, for example, we can add a dataset in hdf5 with the B transposed matrix with : 

```{r hdf5AddtransposedDataset}

# Create dataset BTransposed with B transposed matrix in delayed.hdf5 file at INPUT group
Create_HDF5_matrix(B, "delayed.hdf5", "INPUT", "BTransposed", transp = TRUE);

```


**Open and get hdf5 content file**

We can open an existing file show contents and access data using functions from `rhdf5` package. `rhdf5` is an R interface for HDF5. The file must always be opened before working with it.

With `htls` function we get the the list of an hdf5 content before open the file.

```{r hdf5Show}
# Examine hierarchy before open file
h5ls("delayed.hdf5")
```

Or we can do it after open the file, in that case, we only get the current level in the hierarchical tree

```{r hdf5Open}
# Open file
h5fdelay = H5Fopen("delayed.hdf5")
# Show hdf5 hierarchy (groups)
h5fdelay

```

**Access datasets data**

The \$ operator can be used to access the next group level, this operator reads the object from disk. We can assign the dataset contents to an R object in order to work with it.

```{r hdf5Dataset}
Bdata = h5fdelay$INPUT$B
Bdata[1:3,1:5]
```


Or we can get the data for the transposed version of this matrix created before 
```{r hdf5DatasetTransposed}
BdataTrans = h5fdelay$INPUT$BTransposed
BdataTrans[1:5,1:3]
```

**Close hdf5 file**

After work with hdf5 we always must close the file. We can close only one file with `H5Fclose` function or close all hdf5 opend files with `h5closeAll` function

```{r hdf5Close}
# Close delayed.hdf5 file
H5Fclose(h5fdelay)

# Open 2 files and close all
h5fdelay = H5Fopen("delayed.hdf5")
h5fr = H5Fopen("robjects.hdf5")

h5closeAll()
```

More information about working with hdf5 files in r [@rhdf5]


# Convert text file to HDF5 

Usually we have files with a great amount of data to be loaded in R but in some cases, R can't deal with this data and can't work with this. In order to resolve this issue, you can work directly from data on disk with datasets stored in an HDF5 data file.

You can easily change your text files to HDF5 using the function `Convert_text_to_HDF5`, to do that, we only have to define the input file with the `filename` parameter and the HDF5 destination dataset with parameters `outputfile` for the output file name, `outGroup` with the group to store the dataset and the dataset name with the parameter `outDataset`.

This function **only allows to import numeric data except for rownames and columnames where character are allowed**. 

About the delimiter parameter, the tabulation delimiter is taken in to account by default in function `Convert_text_to_HDF5` but this can be changed easily with `sep` parameter.

In this example we convert the colesterol.csv file and we will save the data to a file 'colesterol_file.hdf5' under the group 'COLESTEROLGROUP' in the dataset 'COLESTEROLDATASET'. In this example the text file contains column names so we set the parameter `header = TRUE`. 

```{r convert_HDF5}

Convert_text_to_HDF5("colesterol.csv", ',', 
                     outputfile = "colesterol_file.hdf5", outGroup = "COLESTEROLGROUP", outDataset = "COLESTEROLDATASET", 
                     header = TRUE)

```

If we observe the new file colesterol_file.hdf5 and its content


```{r read_data_col_HDF5}

h5ls("colesterol_file.hdf5")

# Open 2 files and close all
h5Colesterol = H5Fopen("colesterol_file.hdf5")

# Show hdf5 content dataset
head(h5Colesterol$COLESTEROLGROUP$COLESTEROLDATASET)

# Show colnames 
head(h5Colesterol$COLESTEROLGROUP$.COLESTEROLDATASET_dimnames$`2`)

H5Fclose(h5Colesterol)
```

We can overwrite an existing dataset with new data from 'colesterol_new.csv' file, in that case, we will set the `overwrite` parameter as TRUE `overwrite = TRUE`

```{r overwrite_HDF5}

Convert_text_to_HDF5("colesterol_new.csv", ',', 
                     outputfile = "colesterol_file.hdf5", outGroup = "COLESTEROLGROUP", outDataset = "COLESTEROLDATASET", 
                     header = TRUE, overwrite = TRUE)

```

We can observe that the dataset contains different data


```{r read_new_HDF5}

h5ls("colesterol_file.hdf5")

# Open the hdf5 file
h5Colesterol = H5Fopen("colesterol_file.hdf5")

# Show hdf5 content dataset
head(h5Colesterol$COLESTEROLGROUP$COLESTEROLDATASET)

H5Fclose(h5Colesterol)
```


Now we can operate directly from disk using the datasets in the hdf5 data file


# HDF5 Matrix Multiplication

In this section, different products of matrices and vectors are introduced. The methods implement different strategies including block multiplication algorithms and the use of parallel implementations with hdf5 files datasets. 

`BigDataStatMeth` allows to make an hdf5 matrix multiplication with : 

* two on memory matrices, previously saving the data in a file 
* two matrices inside an hdf5 file.

## Hdf5 Matrix multiplication from data in memory

To work with big matrices `blockmult()` saves matrices in hdf5 file format in order to be able to operate with them in blocks and not overload the memory, by default are considered large matrices if the number of rows or columns is greater than 5000, but it can be changed with `bigmatrix` argument.

In that case, by default the function `blockmult`, by default, this function creates the *tmp_blockmult.hdf5* file with two groups the *INPUT* group with datasets *A* and *B* with original matrices and the *OUTPUT* group with dataset *C* with the results. The file name can be set with `outputfile` parameter.


```{r blockmultbm1}
n <- 1000
p <- 10000
Abig <- matrix(rnorm(n*n), nrow=n, ncol=n)
Bbig <- matrix(rnorm(n*p), nrow=n, ncol=p)


# We want to force it to run in memory
AxBNOBig <- blockmult(Abig, Bbig, bigmatrix = 100000)

# We consider a big matrix if number of rows or columns are > 500
AxBBig3000 <- blockmult(Abig, Bbig, bigmatrix = 500)

```

Depending on whether the calculation has been performed directly in memory or from an hdf5 file, the returned object type is different.

If we work in memory results are returned as a current r matrix object,

```{r blockmultresmat}
class(AxBNOBig)
AxBNOBig[1:5,1:5]
```

if we work in disk, we return a list with the filnme and the dataset that contains the results inside the  hdf5 file. Then we can open the hdf5 file and get the result matrix and the original matrices,

To work in R with hdf5 data object we can use the `rhdf5` function library from `bioconductor`, we can open file and read all the content

```{r blockmultresfile}

h5data <- H5Fopen(AxBBig3000$file)

# We can get wher data is stored
AxBBig3000$dataset

# We observe thad the data is in folder OUTPUT dataset C
reshdf5 <- h5data$OUTPUT$C
reshdf5[1:5,1:5]

all.equal(reshdf5, AxBNOBig)
```

Remember that it is **It is important to close all dataset, group, and file handles when not used** anymore,

```{r blockmultresfileclose}
# Close file
H5Fclose(h5data)
```


## Hdf5 Matrix multiplication from data stored in a file

If we have the matrices stored in hdf5 data file, we can use the function `blockmult_hdf5`. This function allows to perform a matrix multiplication from two matrices stored in a file.

To use the `blockmult_hdf5`, we have to set the parameters `filename` with the file that contains the matrices, `group`, here, we have to indicate under which group are stored the matrices and the datasets names where data is stored parameters `a` and `b`. By default, the results group can be set by `outgroup` parameter by default, the output group is *OUTPUT* .

To show how this function works, we will use the matrices created with `blockmult` function previously used and we will store the results under the 'HDF5_RES' group

```{r blockmult_hdf5_exec}
res <- blockmult_hdf5(filename = 'tmp_blockmult.hdf5', group="INPUT", 
                      a="A", b="B", outgroup = 'HDF5_RES')

# We show the hdf5 content
 h5ls(res$file)
```


Now, we compare the obtained results with `blockmult_hdf5` and `blockmult` functions .
```{r blockmult_hdf5_res}
# Get content
h5data <- H5Fopen(res$file)

# We can get wher data is stored
res$dataset

# We get the results with blockmult (previous step)
resblockmult <- h5data$OUTPUT$C

# And the results obtained from blockmult_hdf5
resblockmult_hdf5 <- h5data$HDF5_RES$A_x_B
resblockmult_hdf5[1:5,1:5]

# Close the file
h5closeAll()

all.equal(resblockmult, resblockmult_hdf5)
```


We obtain the same result using `blockmult` and `blockmult_hdf5`.


# Cross-product and Transposed Cross-product with hdf5 files

To perform a cross-product $C = A^t A$ you can call `Crossprod_hdf5()` function. We just have to set the parameters `filename`, `group` and `A` with the file, group and the dataset that we want to use and optionally the `outgroup` parameter to set the output group where we want to store results.

The result dataset is stored with *CrossProd_<MatrixName>x<MatrixName>* pattern.


```{r crossprod }
res <- Crossprod_hdf5(filename = 'robjects.hdf5', group="INPUT", A="A", outgroup = "RESULTS")
```

We obtain the expected values computed using `crossprod` function.

```{r check_cp}

# Get content
h5data <- H5Fopen(res$file)

# We can get wher data is stored
res$dataset

# We get the Crossprod Results and the original matrix
resCProd <- h5data$RESULTS$CrossProd_AxA
A <- h5data$INPUT$A

# Close the file
h5closeAll()

# Show results
resCProd[1:5,1:5]

all.equal(resCProd, crossprod(A))
```

You may also compute the Crossprod with two different matrix $C = A^t B$ , 

```{r crossprod_ab }

set.seed(5234)
n <- 500
m <- 600
B <- matrix(rnorm(n*m,mean=0,sd=1), n,m)

Create_HDF5_matrix(B, filename = 'robjects.hdf5', group="INPUT", dataset = "B2")

# Get Crossprod with two matrix
res <- Crossprod_hdf5(filename = 'robjects.hdf5', group="INPUT", A="A", 
                      groupB = "INPUT", B = "B2", outgroup = "RESULTS")

```

We obtain the expected values computed using `crossprod` function.

```{r check_cp_ab}

# Get content
h5data <- H5Fopen(res$file)

# We can get wher data is stored
res$dataset

# We get the Crossprod Results and the original matrix
resCProd <- h5data$RESULTS$CrossProd_AxB2
A <- h5data$INPUT$A
B <- h5data$INPUT$B2

# Close the file
h5closeAll()

# Show results
resCProd[1:5,1:5]

all.equal(resCProd, crossprod(A,B))
```



you may also set use `tCrossprod_hdf5()` function to get a transposed cross-product $C = A A^t$ . In that case the result dataset is stored with *tCrossProd_<MatrixName>x<MatrixName>* pattern.


```{r tcrossprod }
res <- tCrossprod_hdf5(filename = 'robjects.hdf5', group="INPUT", A="A", outgroup = "RESULTS")
```

We obtain the expected values computed using `crossprod` function.

```{r check_tcp}

# Get content
h5data <- H5Fopen(res$file)

# We can get wher data is stored
res$dataset

# We get the Crossprod Results and the original matrix
restCProd <- h5data$RESULTS$tCrossProd_AxA
A <- h5data$INPUT$A

# Close the file
h5closeAll()

# Show results
restCProd[1:5,1:5]

all.equal(restCProd, tcrossprod(A))
```

You may also compute the tCrossprod with two different matrix $C = A B^t$ 

```{r tcrossprod_ab }

set.seed(5234)
n <- 500
m <- 600
B <- matrix(rnorm(n*m,mean=0,sd=1), n,m)

Create_HDF5_matrix(B, filename = 'robjects.hdf5', group="INPUT", dataset = "B3")

# Get Crossprod with two matrix
res <- tCrossprod_hdf5(filename = 'robjects.hdf5', group="INPUT", A="A", 
                       groupB = "INPUT", B = "B3", outgroup = "RESULTS")

```

We obtain the expected values computed using `crossprod` function.

```{r check_tcp_ab}

# Get content
h5data <- H5Fopen(res$file)

# We can get wher data is stored
res$dataset

# We get the Crossprod Results and the original matrix
restCProd <- h5data$RESULTS$tCrossProd_AxB3
A <- h5data$INPUT$A
B <- h5data$INPUT$B3

# Close the file
h5closeAll()

# Show results
restCProd[1:5,1:5]

all.equal(restCProd, tcrossprod(A,B))
```



# Singular Value Decomposition (SVD)

The SVD of an $m \times n$ real or complex matrix $A$ is a factorization of the form:

$$U\Sigma { V }^{ T }$$

where :

-$U$ is a $m \times m$ real or complex unitary matrix
-$\Sigma$ is a $m \times n$ rectangular diagonal matrix with non-negative real numbers on the diagonal
-$V$ is a $n \times n$ real or complex unitary matrix.

Notice that:

- The diagonal entries $\sigma_i$ of $\Sigma$ are known as the singular values of $A$.
- The columns of $U$ are called the left-singular vectors of $A$.
- The columns of $V$ are called the right-singular vectors of $A$.


## Block Singular Values Decomposition

This method was developed in June 2016 by M. A. Iwen and B. W. Ong. The authors describe it as a distributed and incremental SVD algorithm that is useful for agglomerative data analysis on large networks. The algorithm calculates the singular values and left singular vectors of a matrix A, by first, partitioning it by columns. This creates a set of submatrices of A with the same number of rows, but only some of its columns. After that, the SVD of each of the submatrices is computed. The final step consists of combining the results obtained by merging them again and computing the SVD of the resulting matrix.

```{r BSVDImg, out.width = '100%', fig.align = 'center', fig.cap = "Flowchart for a two-level hierarchical Block SVD algorithm", echo=FALSE}
knitr::include_graphics("imgs/blocksvd.png")
```

This method is implemented in `bdSVD_hdf5` function, this function works directly on hdf5 data format, loading in memory only the data to perform calculations and saving the results again in the hdf5 file for later use.

We have to indicate the file to work with, the dataset name and the group under the dataset is located :

```{r BlockSVDNorm}

# Create dataframe data with Ad matrix in delayed.hdf5 file at OMIC group
set.seed(5234)
n <- 150000
m <- 50
odata <- matrix(rnorm(n*m,mean=0,sd=1), n,m)

Create_HDF5_matrix(odata, "delayed.hdf5", "OMICS", "data")

# Direct from hdf5 data file
svdh5 <- bdSVD_hdf5("delayed.hdf5", "OMICS", "data")

# with R implementation from data in memory
fprova <- H5Fopen("delayed.hdf5")
  # get results svd (d)
  svd_hdf5_d <- fprova$SVD$data$d[1:7] 
  # get data
  omdata <- fprova$OMICS$data 
h5closeAll()

# Results in hdf5 file for d
svd_hdf5_d[1:7]

svd <- svd(scale(omdata))
svd$d[1:7]


```


Like in Simple Singular Value Decomposition we can normalize, center or scale data before proceed with SVD decomposition with `bscale` and `bcenter` parameters, by default this parameter are TRUE, data is normalized before SVD decomposition. To proceed with SVD without normalization :

```{r BlockSVDNotNorm}
# Direct from hdf5 data file
svdh5 <- bdSVD_hdf5("delayed.hdf5", "OMICS", "data", 
                    bcenter = FALSE, bscale = FALSE)

# get results svd (d)
fprova <- H5Fopen("delayed.hdf5")
  svd_hdf5_d <- fprova$SVD$data$d[1:7] 
h5closeAll()

# SVD (d) from file - data not normalized
svd_hdf5_d

# with R implementation from data in memory
svd <- svd(omdata)
svd$d[1:7]

```

In the SVD decomposition by blocks we can indicate the number of decomposition levels and number of local SVDs to concatenate at each level with parameters `q` and `k` respectively, by default `q = 1` one level with `k=2`.

```{r BlockSVDk4}
# Block decomposition with 1 level and 4 local SVDs at each level
svdh5 <- bdSVD_hdf5("delayed.hdf5", "OMICS", "data", q=1, k=4 )


# get results svd (d)
fprova <- H5Fopen("delayed.hdf5")
  svd_hdf5_d <- fprova$SVD$data$d[1:7] 
h5closeAll()

# SVD (d) from file - data not normalized
svd_hdf5_d

# with R implementation from data in memory
svd <- svd(scale(omdata))
svd$d[1:7]

```



<!-- # Using algebra procedure for `DelayedMatrix` objects to implement basic statistical methods -->


<!-- ## Principal component analysis (PCA) -->

<!-- Let us illustrate how to perform a PCA using miRNA data obtained from TCGA corresponding to 3 different tumors: melanoma (ME), leukemia (LEU) and centran nervous system (CNS). Data is available at `BigDataStatMeth` and can be loaded by simply: -->

<!-- ```{r mirNA} -->
<!-- data(miRNA) -->
<!-- data(cancer) -->
<!-- dim(miRNA) -->

<!-- Create_HDF5_matrix(odata, "delayed.hdf5", "OMICS", "miRNA") -->

<!-- ``` -->

<!-- We observe that there are a total of 21 individuals and 537 miRNAs. The vector `cancer` contains the type of tumor of each individual. For each type we have: -->

<!-- ```{r tab} -->
<!-- table(cancer) -->
<!-- ``` -->

<!-- Now, the typical principal component analysis on the samples -->
<!-- can be run on the `miRNA` matrix since it has miRNAs in columns and individuals in rows -->

<!-- ```{r pca} -->
<!-- pc <- prcomp(miRNA) -->
<!-- ``` -->

<!-- We can plot the two first components with: -->

<!-- ```{r plot_pca} -->
<!-- plot(pc$x[, 1], pc$x[, 2], -->
<!--      main = "miRNA data on tumor samples", -->
<!--      xlab = "PC1", ylab = "PC2", type="n") -->
<!-- abline(h=0, v=0, lty=2) -->
<!-- points(pc$x[, 1], pc$x[, 2], col = cancer, -->
<!--        pch=16, cex=1.2) -->
<!-- legend("topright", levels(cancer), pch=16, col=1:3) -->
<!-- ``` -->

<!-- The same analysis can be performed using SVD decomposition and having miRNAs as a `DelayedMatrix` object. The PCA is equivalent to performing the SVD on the centered data, where the centering occurs on the columns. In that case the function `bdSVD` requires to set the argument `bcenter` and `bscale` equal to TRUE, the dafault values. Notice that `sweep()` and `colMeans()` functions can be applied to a `DelayedMAtrix` object since this method is implemented for that type of objects in the `DelayedArray` package. -->

<!-- ```{r cia_da} -->
<!-- hdf5PCA <- bdPCA_hdf5("delayed.hdf5", group = "OMICS", dataset= "miRNA", bcenter = FALSE, bscale = FALSE) -->


<!-- ``` -->


<!-- The PCA plot for the two principal components can then be be obtained by: -->

<!-- ```{r plot_svd_da} -->
<!-- plot(svd.da$u[, 1], svd.da$u[, 2], -->
<!--      main = "miRNA data on tumor samples", -->
<!--      xlab = "PC1", ylab = "PC2", type="n") -->
<!-- abline(h=0, v=0, lty=2) -->
<!-- points(svd.da$u[, 1], svd.da$u[, 2], col = cancer, -->
<!--        pch=16, cex=1.2) -->
<!-- legend("topright", levels(cancer), pch=16, col=1:3) -->
<!-- ``` -->

<!-- We can observe that both figures are equal irrespective to a sign change of second component (that can happen in SVD). -->

<!-- # Statistical methods implemented for `DelayedMatrix` objects -->

<!-- In this section we illustrate how to estimate some of the state of the art methods used in omic data analyses having a `DelayedMatrix` as the input object. So far, we have implemented the lasso regression, but other methods are going to be implemented in the near future. -->


# Utils for data analysis working directly in hdf5 files

## Data imputation

Imputation in genetics refers to the statistical inference of unobserved genotypes. In genetic epidemiology and quantitative genetics, researchers aim at identifying genomic locations where variation between individuals is associated with variation in traits of interest between individuals.

At the moment BigDataStatMeth has implemented snps imputation following the encoding used in Genomic Data Structure file format (GDS files) where SNPs are codded as 0 : two B alleles, 1 : one A allele and one B allele, 2 : two A alleles and 3 : missing data. BigDataStatMeth impute data where SNPs values are '3' (missing data).

The imputation method implemented is generated from the information of the SNP data that contains the missing data, it is performed by generating a random value following a discrete distribution. If in the SNP we find 70% of '0', 25% of '1' and 5% of '2', the value of the missing data will be imputed with a probability of 70% to be '0', 25% to be '1' and 5% to be '2'.

We first simulate a genotype matrix with 0, 1, 2 and 3 and store it in hdf5 file and show data stored in file:
```{r snpgenotype}
set.seed(108432)
geno.sim <- matrix(sample(0:3, 10000, replace = TRUE), byrow = TRUE, ncol = 10)
Create_HDF5_matrix(geno.sim, "delayed.hdf5", "OMICS", "geno")

# Get data and show the first 5 rows
h5fsvd = H5Fopen("delayed.hdf5")
geno <- h5fsvd$OMICS$geno
h5closeAll()

geno[1:5,1:10]

```

Remember, we allways have to close hdf5 data file. Now we impute values where genotype = '3' (missing data in GDS format). To apply imputation we only have to indicate the file where dataset is, the group and the dataset name, optionally, we indicate were we want to store results with `outputgroup` and `outputfolder`, by default the results are stored in the input dataset.
An important parameter is `bycols`, with `bycols` we can inform if we want to impute data by cols `bycols = TRUE` or by rows `bycols = FALSE`, the default imputation implemented method is by cols.

In next example we perform the imputation in "imputedgeno" dataset inside the group "OMICS" and the imputation method applied takes into account the probabilities in columns (default).
```{r impute}
bdImputeSNPHDF5("delayed.hdf5", group="OMICS", dataset="geno",
                outgroup="OMICS", outdataset="imputedgeno")
```

Now we get the imputation results

```{r impute2}
# Get imputed data and show the first 5 rows
h5fsvd = H5Fopen("delayed.hdf5")
imputedgeno <- h5fsvd$OMICS$imputedgeno
h5closeAll()

imputedgeno[1:5,1:10]
```


## Normalization, center and scale

Normalization is one of the most important procedures in genomics data analysis. A typical dataset contains more than one sample and we are almost always interested in making comparisons between these and in order to make comparisons we need to normalize data.

In BigDataStatMeth we implemented a normalization method that works directly with datasets stored in hdf5 file format, like in other functions implemented in BigDataSet that works with hdf5 data files, we have to indicate where the data is, the filename, the name of the group under the dataset is and the dataset name. In normalization function, we can indicate if we want to center data, scale or center and scale (default option). The applied formula to normalize data by default in implemented function is :
$$ \frac{X - \mu}{\sigma}$$

The implemented function to normalize data directo from hdf5 data file is `Normalize_hdf5`.

To show you an example, we will use the imputed geno dataset *imputedgeno* created before in imputation method example. To normalize data we have to indicate where the dataset is stored and we have two optiona paramters `bcenter` and `bscale` with default value = TRUE. In order to normalize data, we don't need to modify these optional parameters. If we only want to center data, then we have to put `bscale = FALSE` or if we only want to scale data `bcenter` parameter must be FALSE.

**Important : ** The normalization results are stored under "NORMALIZED" group, under this group, we found all normalized datasets with the same structure under root, for example, if we normalize dataset genoimputed under OMICS group, the normalized results will be in "NORMALIZED"/"OMICS"/genoimputed

Let show an example normalizing imputed geno under group "OMICS"  in delayed.hdf5 file:

```{r normalize}

Normalize_hdf5("delayed.hdf5", group="OMICS", dataset="imputedgeno")
```

The results will be under groups "NORMALIZED" - "OMICS" in dataset genoimputed

```{r normalizeres}
# Geno after normalization
h5fsvd = H5Fopen("delayed.hdf5")
genonormalized <- h5fsvd$NORMALIZED$OMICS$geno
h5closeAll()

genonormalized[1:5,1:10]
```


## QC - Remove low data

Missing data in meta-analysis of genetic association studies are unavoidable. In BigDataStatMeth we implemented a function to remove those SNPs with a hight percentage of missing data. The implemented function to remove SNPs wit a certain percentage of missing data is the function `bdRemovelowdata`

Like in imputation method, in remove low data method we have to inform the input data indicating  the filename, the group and the dataset name and we also have to inform the dataset were data will be stored after removing the SNPs with parameters `outgroup` and `outdataset`.

To remove low data we have to infrorm if we want to remove data by cols `SNPincols = TRUE` or by rows `SNPincols = FALSE` and the percentage of missing data `pcent` parameter. `pcent` of missing data refers to missing percentage in columns if we defined `SNPincols = TRUE` or `pcent` refers to rows if `SNPincols = FALSE`.

To show how renove low data works we use the previous dataset stored in *delayed.hdf5* file, in that case, we will assume that SNPs are in columns and we will remove those SNPs where missing data is greater than 40%.

```{r removelow}
bdRemovelowdata("delayed.hdf5", group="OMICS", dataset="geno",
                outgroup="OMICS", outdataset="removedgeno",
                SNPincols = TRUE, pcent = 0.39)
```

After remove the SNPs with pcent greater than 39% results have been saved in dataset "*removedgeno*" under group "*OMICS*", we observe the resulting dataset

```{r removelow2}
# Get imputed data and show the first 5 rows
h5fsvd = H5Fopen("delayed.hdf5")
removedgeno <- h5fsvd$OMICS$removedgeno
h5closeAll()

removedgeno[1:5,1:10]
```



# References

<a id="rhdf5">[1]</a>
Fischer, Bernd, Gregoire Pau, and Mike Smith. (2019).
R Interface to Hdf5.
https://github.com/grimbough/rhdf5

# Session information

```{r sesinfo }
sessionInfo()
```

