<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>What is HDF5? ‚Äì BigDataStatMeth</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/favicon.png" rel="icon" type="image/png">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-065a5179aebd64318d7ea99d77b64a9e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-69e22614feb0789a4b1e69368fd0350a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<style>html{ scroll-behavior: smooth; }</style>


<link rel="stylesheet" href="../../assets/custom.css">
</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">BigDataStatMeth</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../get-started/index.html"> 
<span class="menu-text">Get Started</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-learn" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Learn</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-learn">    
        <li class="dropdown-header">‚îÄ‚îÄ‚îÄ Theory ‚îÄ‚îÄ‚îÄ</li>
        <li>
    <a class="dropdown-item" href="../../fundamentals/index.html">
 <span class="dropdown-text">Fundamentals</span></a>
  </li>  
        <li class="dropdown-header">‚îÄ‚îÄ‚îÄ Practice ‚îÄ‚îÄ‚îÄ</li>
        <li>
    <a class="dropdown-item" href="../../tutorials/index.html">
 <span class="dropdown-text">Tutorials</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../practical-workflows/index.html">
 <span class="dropdown-text">Practical Workflows</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-api-reference" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">API Reference</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-api-reference">    
        <li>
    <a class="dropdown-item" href="../../api-reference/r/index.html">
 <span class="dropdown-text">R Functions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../api-reference/cpp/index.html">
 <span class="dropdown-text">C++ API</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-developers" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Developers</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-developers">    
        <li>
    <a class="dropdown-item" href="../../developer-guide/index.html">
 <span class="dropdown-text">Developer Guide</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../developer-guide/contributing/index.qmd">
 <span class="dropdown-text">Contributing</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-resources" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Resources</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-resources">    
        <li>
    <a class="dropdown-item" href="../../best-practices/index.html">
 <span class="dropdown-text">Best Practices</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../resources/index.html">
 <span class="dropdown-text">Additional Resources</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/isglobal-brge/BigDataStatMeth"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../fundamentals/hdf5/index.html">HDF5 Data Format</a></li><li class="breadcrumb-item"><a href="../../fundamentals/hdf5/what-is-hdf5.html">What is HDF5?</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../../index.html" class="sidebar-logo-link">
      </a>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Fundamentals</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">HDF5 Data Format</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/hdf5/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">HDF5 Data Format</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/hdf5/what-is-hdf5.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">What is HDF5?</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/hdf5/structure.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">HDF5 Structure and Organization</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/hdf5/operations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">HDF5 Operations</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/hdf5/performance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">HDF5 Performance Optimization</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/hdf5/best-practices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">HDF5 Best Practices</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Linear Algebra</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/linear-algebra/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Linear Algebra</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Statistical Methods</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/statistical-methods/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Statistical Methods</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Big Data Concepts</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../fundamentals/big-data-concepts/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Big Data Concepts</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#quick-overview" id="toc-quick-overview" class="nav-link active" data-scroll-target="#quick-overview">Quick Overview</a></li>
  <li><a href="#conceptual-understanding" id="toc-conceptual-understanding" class="nav-link" data-scroll-target="#conceptual-understanding">Conceptual Understanding</a>
  <ul class="collapse">
  <li><a href="#what-is-hdf5" id="toc-what-is-hdf5" class="nav-link" data-scroll-target="#what-is-hdf5">What is HDF5?</a></li>
  <li><a href="#why-use-hdf5-for-big-data" id="toc-why-use-hdf5-for-big-data" class="nav-link" data-scroll-target="#why-use-hdf5-for-big-data">Why Use HDF5 for Big Data?</a></li>
  <li><a href="#the-hierarchy-concept" id="toc-the-hierarchy-concept" class="nav-link" data-scroll-target="#the-hierarchy-concept">The Hierarchy Concept</a></li>
  <li><a href="#real-world-analogy" id="toc-real-world-analogy" class="nav-link" data-scroll-target="#real-world-analogy">Real-World Analogy</a></li>
  </ul></li>
  <li><a href="#technical-details" id="toc-technical-details" class="nav-link" data-scroll-target="#technical-details">Technical Details</a></li>
  <li><a href="#using-hdf5-in-r" id="toc-using-hdf5-in-r" class="nav-link" data-scroll-target="#using-hdf5-in-r">Using HDF5 in R</a>
  <ul class="collapse">
  <li><a href="#available-packages" id="toc-available-packages" class="nav-link" data-scroll-target="#available-packages">Available Packages</a></li>
  <li><a href="#basic-operations-with-rhdf5" id="toc-basic-operations-with-rhdf5" class="nav-link" data-scroll-target="#basic-operations-with-rhdf5">Basic Operations with rhdf5</a></li>
  <li><a href="#simplified-operations-with-bigdatastatmeth" id="toc-simplified-operations-with-bigdatastatmeth" class="nav-link" data-scroll-target="#simplified-operations-with-bigdatastatmeth">Simplified Operations with BigDataStatMeth</a></li>
  <li><a href="#best-practices" id="toc-best-practices" class="nav-link" data-scroll-target="#best-practices">Best Practices</a></li>
  <li><a href="#common-pitfalls-and-solutions" id="toc-common-pitfalls-and-solutions" class="nav-link" data-scroll-target="#common-pitfalls-and-solutions">Common Pitfalls and Solutions</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources">Additional Resources</a>
  <ul class="collapse">
  <li><a href="#official-documentation" id="toc-official-documentation" class="nav-link" data-scroll-target="#official-documentation">Official Documentation</a></li>
  <li><a href="#tutorials" id="toc-tutorials" class="nav-link" data-scroll-target="#tutorials">Tutorials</a></li>
  <li><a href="#tools" id="toc-tools" class="nav-link" data-scroll-target="#tools">Tools</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../fundamentals/hdf5/index.html">HDF5 Data Format</a></li><li class="breadcrumb-item"><a href="../../fundamentals/hdf5/what-is-hdf5.html">What is HDF5?</a></li></ol></nav>
<div class="quarto-title">
<h1 class="title">What is HDF5?</h1>
<p class="subtitle lead">Understanding Hierarchical Data Format</p>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="content-guide">
<p><strong>Choose your path:</strong></p>
<ul>
<li>üéØ <a href="#quick-overview">Quick Overview</a> - 2 min read</li>
<li>üí° <a href="#conceptual-understanding">Conceptual Understanding</a> - 10 min read</li>
<li>üîß <a href="#technical-details">Technical Details</a> - 15 min read (advanced)</li>
<li>üíª <a href="#using-hdf5-in-r">Using HDF5 in R</a> - 10 min read</li>
</ul>
</div>
<hr>
<section id="quick-overview" class="level2">
<h2 class="anchored" data-anchor-id="quick-overview">Quick Overview</h2>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Note</span>üìã TL;DR - What you need to know
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>HDF5 (Hierarchical Data Format version 5)</strong> is a file format and technology suite designed to store and organize large amounts of data efficiently.</p>
<p><strong>Key advantages:</strong> - <strong>Scalable</strong>: Handles datasets from kilobytes to petabytes - <strong>Fast</strong>: Optimized I/O operations for big data - <strong>Flexible</strong>: Stores any type of data in a single file - <strong>Portable</strong>: Works across different operating systems and platforms - <strong>Self-describing</strong>: Metadata embedded in the file</p>
<p><strong>When to use HDF5:</strong> - Working with datasets too large to fit in memory - Need fast access to subsets of large matrices - Storing multiple related datasets together - Sharing data across different platforms</p>
<p><strong>In BigDataStatMeth:</strong> All matrix operations can work directly on HDF5 files without loading data into memory.</p>
</div>
</div>
<hr>
</section>
<section id="conceptual-understanding" class="level2">
<h2 class="anchored" data-anchor-id="conceptual-understanding">Conceptual Understanding</h2>
<section id="what-is-hdf5" class="level3">
<h3 class="anchored" data-anchor-id="what-is-hdf5">What is HDF5?</h3>
<p>Think of HDF5 as a <strong>sophisticated filing cabinet</strong> for your data. Just as a filing cabinet has drawers (groups) that contain folders (more groups) with documents (datasets) inside, HDF5 organizes data in a hierarchical structure.</p>
<p><strong>Traditional approach:</strong></p>
<pre><code>project/
‚îú‚îÄ‚îÄ gene_expression.csv      (5 GB)
‚îú‚îÄ‚îÄ methylation.csv          (8 GB)  
‚îú‚îÄ‚îÄ clinical_data.csv        (2 MB)
‚îî‚îÄ‚îÄ metadata.txt</code></pre>
<p><strong>HDF5 approach:</strong></p>
<pre><code>project.hdf5
‚îú‚îÄ‚îÄ /omics/
‚îÇ   ‚îú‚îÄ‚îÄ gene_expression      (5 GB dataset)
‚îÇ   ‚îî‚îÄ‚îÄ methylation          (8 GB dataset)
‚îú‚îÄ‚îÄ /clinical/
‚îÇ   ‚îî‚îÄ‚îÄ patient_data         (2 MB dataset)
‚îî‚îÄ‚îÄ attributes: metadata</code></pre>
<p>All in <strong>one file</strong>, with <strong>fast access</strong> to any part without reading everything.</p>
</section>
<section id="why-use-hdf5-for-big-data" class="level3">
<h3 class="anchored" data-anchor-id="why-use-hdf5-for-big-data">Why Use HDF5 for Big Data?</h3>
<section id="out-of-core-computation" class="level4">
<h4 class="anchored" data-anchor-id="out-of-core-computation">1. <strong>Out-of-Core Computation</strong></h4>
<p>Imagine you have a <strong>100 GB matrix</strong> of genomic data. Traditional R would require:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Traditional approach - CRASHES!</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"huge_data.csv"</span>)  <span class="co"># Tries to load 100 GB into RAM</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">cor</span>(data)                 <span class="co"># Out of memory!</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<p>With HDF5 and BigDataStatMeth:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># HDF5 approach - Works!</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">bdCreate_hdf5_matrix</span>(<span class="st">"data.hdf5"</span>, data, <span class="st">"analysis"</span>, <span class="st">"genotypes"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">bdCrossprod_hdf5</span>(<span class="st">"data.hdf5"</span>, <span class="st">"analysis"</span>, <span class="st">"genotypes"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Only loads small chunks at a time</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="partial-reading" class="level4">
<h4 class="anchored" data-anchor-id="partial-reading">2. <strong>Partial Reading</strong></h4>
<p>Need only 100 rows from a million-row dataset? HDF5 reads <strong>only what you need</strong>:</p>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read specific rows without loading entire dataset</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>h5file <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(<span class="st">"data.hdf5"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>subset <span class="ot">&lt;-</span> h5file<span class="sc">$</span>data[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, ]  <span class="co"># Reads only 100 rows</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="performance-optimizations" class="level4">
<h4 class="anchored" data-anchor-id="performance-optimizations">3. <strong>Performance Optimizations</strong></h4>
<p>HDF5 includes built-in optimizations:</p>
<ul>
<li><strong>Chunking</strong>: Data stored in optimal-sized blocks for fast access</li>
<li><strong>Compression</strong>: Reduce file size (gzip, LZF, etc.)</li>
<li><strong>Parallel I/O</strong>: Multiple processes can read simultaneously</li>
<li><strong>Caching</strong>: Frequently accessed data cached in memory</li>
</ul>
</section>
</section>
<section id="the-hierarchy-concept" class="level3">
<h3 class="anchored" data-anchor-id="the-hierarchy-concept">The Hierarchy Concept</h3>
<p>HDF5‚Äôs power comes from its <strong>tree-like structure</strong>, similar to a file system:</p>
<pre><code>data.hdf5
‚îú‚îÄ‚îÄ /raw_data/
‚îÇ   ‚îú‚îÄ‚îÄ samples_x_snps          [2,000 √ó 500,000 matrix]
‚îÇ   ‚îî‚îÄ‚îÄ phenotypes              [2,000 √ó 50 matrix]
‚îú‚îÄ‚îÄ /processed/
‚îÇ   ‚îú‚îÄ‚îÄ qc_filtered             [1,800 √ó 450,000 matrix]
‚îÇ   ‚îú‚îÄ‚îÄ imputed                 [1,800 √ó 450,000 matrix]
‚îÇ   ‚îî‚îÄ‚îÄ normalized              [1,800 √ó 450,000 matrix]
‚îî‚îÄ‚îÄ /results/
    ‚îú‚îÄ‚îÄ pca_components          [1,800 √ó 50 matrix]
    ‚îú‚îÄ‚îÄ eigenvalues             [50 vector]
    ‚îî‚îÄ‚îÄ loadings                [450,000 √ó 50 matrix]</code></pre>
<p><strong>Benefits:</strong> - All analysis stages in one file - Organized workflow - Easy to track processing steps - No file management headaches</p>
</section>
<section id="real-world-analogy" class="level3">
<h3 class="anchored" data-anchor-id="real-world-analogy">Real-World Analogy</h3>
<p><strong>Library (Traditional files):</strong> - Each book (dataset) is a separate file - To read, must carry entire book to desk (load into RAM) - Can only work with books that fit on your desk (RAM limit)</p>
<p><strong>Digital Library with Smart Search (HDF5):</strong> - All books in one database - Search finds exactly what you need - Read only the pages you want - Instant access to any section - Desk size doesn‚Äôt matter</p>
<hr>
</section>
</section>
<section id="technical-details" class="level2">
<h2 class="anchored" data-anchor-id="technical-details">Technical Details</h2>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>üî¢ Technical Specifications (click to expand)
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="file-structure" class="level3">
<h3 class="anchored" data-anchor-id="file-structure">File Structure</h3>
<p>HDF5 organizes data using two primary object types:</p>
<section id="groups" class="level4">
<h4 class="anchored" data-anchor-id="groups">1. Groups</h4>
<ul>
<li>Container structures (like directories)</li>
<li>Can contain datasets and other groups</li>
<li>Form the hierarchical tree</li>
<li>Identified by paths: <code>/group/subgroup</code></li>
</ul>
</section>
<section id="datasets" class="level4">
<h4 class="anchored" data-anchor-id="datasets">2. Datasets</h4>
<ul>
<li>Multidimensional arrays of homogeneous data</li>
<li>Can be of any dimension (0D scalar to nD arrays)</li>
<li>Support chunking and compression</li>
<li>Can have attached metadata (attributes)</li>
</ul>
</section>
</section>
<section id="storage-layout" class="level3">
<h3 class="anchored" data-anchor-id="storage-layout">Storage Layout</h3>
<p>HDF5 offers three storage layouts:</p>
<p><strong>Contiguous Layout:</strong></p>
<pre><code>[1 2 3 4 5 6 7 8 9 10 11 12]  &lt;- Continuous block</code></pre>
<ul>
<li>Fastest for sequential access</li>
<li>No overhead</li>
<li>Entire dataset must fit in one block</li>
</ul>
<p><strong>Chunked Layout:</strong></p>
<pre><code>[1 2 3 4] [5 6 7 8] [9 10 11 12]  &lt;- Independent chunks</code></pre>
<ul>
<li>Enables compression</li>
<li>Allows partial I/O</li>
<li>Extensible datasets</li>
<li>Optimal for big data</li>
</ul>
<p><strong>Compact Layout:</strong></p>
<pre><code>Header: [1 2 3 4 5]  &lt;- Stored in header</code></pre>
<ul>
<li>For small datasets (&lt; 64 KB)</li>
<li>Minimal overhead</li>
<li>Fastest access</li>
</ul>
</section>
<section id="data-types" class="level3">
<h3 class="anchored" data-anchor-id="data-types">Data Types</h3>
<p>HDF5 supports comprehensive data types:</p>
<ul>
<li><strong>Integer</strong>: 8, 16, 32, 64 bit (signed/unsigned)</li>
<li><strong>Float</strong>: 32, 64 bit (IEEE 754)</li>
<li><strong>String</strong>: Fixed or variable length</li>
<li><strong>Compound</strong>: Structures with named fields</li>
<li><strong>Array</strong>: Fixed-size multidimensional arrays</li>
<li><strong>Enum</strong>: Enumerated types</li>
<li><strong>Bitfield</strong>: Packed bit sequences</li>
</ul>
</section>
<section id="chunking-strategy" class="level3">
<h3 class="anchored" data-anchor-id="chunking-strategy">Chunking Strategy</h3>
<p>Chunks are the <strong>fundamental I/O unit</strong>:</p>
<pre><code>Matrix [1000 √ó 1000]
‚îú‚îÄ‚îÄ Chunk [100 √ó 100]  (100 chunks total)
‚îú‚îÄ‚îÄ Chunk [100 √ó 100]
‚îî‚îÄ‚îÄ ...</code></pre>
<p><strong>Optimal chunk size:</strong> - Not too small: Minimize overhead - Not too large: Enable partial I/O - <strong>Rule of thumb</strong>: 10 KB to 1 MB per chunk</p>
</section>
<section id="compression" class="level3">
<h3 class="anchored" data-anchor-id="compression">Compression</h3>
<p>Supported filters:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th>Filter</th>
<th>Ratio</th>
<th>Speed</th>
<th>Use Case</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>GZIP</strong> (levels 1-9)</td>
<td>High</td>
<td>Slow</td>
<td>Maximum compression</td>
</tr>
<tr class="even">
<td><strong>LZF</strong></td>
<td>Medium</td>
<td>Fast</td>
<td>Balanced</td>
</tr>
<tr class="odd">
<td><strong>SZIP</strong></td>
<td>Medium</td>
<td>Fast</td>
<td>Scientific data</td>
</tr>
<tr class="even">
<td><strong>Blosc</strong></td>
<td>High</td>
<td>Very Fast</td>
<td>Numerical arrays</td>
</tr>
</tbody>
</table>
<p><strong>Example compression ratios:</strong> - Random data: ~1.1x (poor) - Genomic data: ~3-5x (good) - Sparse matrices: ~10-50x (excellent)</p>
</section>
<section id="parallel-io" class="level3">
<h3 class="anchored" data-anchor-id="parallel-io">Parallel I/O</h3>
<p>HDF5 supports parallel operations:</p>
<ul>
<li><strong>MPI-IO</strong>: Parallel reads/writes</li>
<li><strong>Independent I/O</strong>: Each process accesses different regions</li>
<li><strong>Collective I/O</strong>: Coordinated access for optimization</li>
</ul>
<p><strong>Requirement:</strong> HDF5 compiled with parallel support</p>
</section>
<section id="file-overhead" class="level3">
<h3 class="anchored" data-anchor-id="file-overhead">File Overhead</h3>
<p>HDF5 has minimal overhead:</p>
<ul>
<li><strong>Superblock</strong>: ~512 bytes (file metadata)</li>
<li><strong>B-tree nodes</strong>: ~1 KB per group</li>
<li><strong>Chunk index</strong>: Depends on number of chunks</li>
<li><strong>Attributes</strong>: Small (typically &lt; 1 KB each)</li>
</ul>
<p>For large datasets, overhead is <strong>negligible</strong> (&lt; 0.1% of file size).</p>
</section>
</div>
</div>
</div>
<hr>
</section>
<section id="using-hdf5-in-r" class="level2">
<h2 class="anchored" data-anchor-id="using-hdf5-in-r">Using HDF5 in R</h2>
<section id="available-packages" class="level3">
<h3 class="anchored" data-anchor-id="available-packages">Available Packages</h3>
<p>Three main packages for HDF5 in R:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 32%">
<col style="width: 32%">
<col style="width: 35%">
</colgroup>
<thead>
<tr class="header">
<th>Package</th>
<th>Purpose</th>
<th>Best For</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>rhdf5</strong></td>
<td>Low-level HDF5 operations</td>
<td>File management, inspection</td>
</tr>
<tr class="even">
<td><strong>HDF5Array</strong></td>
<td>Bioconductor integration</td>
<td>Genomic data, SummarizedExperiment</td>
</tr>
<tr class="odd">
<td><strong>BigDataStatMeth</strong></td>
<td>Statistical methods on HDF5</td>
<td>Matrix algebra, PCA, regression</td>
</tr>
</tbody>
</table>
</section>
<section id="basic-operations-with-rhdf5" class="level3">
<h3 class="anchored" data-anchor-id="basic-operations-with-rhdf5">Basic Operations with rhdf5</h3>
<section id="creating-an-hdf5-file" class="level4">
<h4 class="anchored" data-anchor-id="creating-an-hdf5-file">Creating an HDF5 file</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rhdf5)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create file with group and dataset</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="fu">h5createFile</span>(<span class="st">"mydata.hdf5"</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">h5createGroup</span>(<span class="st">"mydata.hdf5"</span>, <span class="st">"experiment"</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="fu">h5createDataset</span>(<span class="st">"mydata.hdf5"</span>, <span class="st">"experiment/data"</span>, </span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">1000</span>, <span class="dv">500</span>), </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">chunk =</span> <span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">100</span>),</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">level =</span> <span class="dv">6</span>)  <span class="co"># gzip compression level</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Write data</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>matrix_data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">500000</span>), <span class="dv">1000</span>, <span class="dv">500</span>)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a><span class="fu">h5write</span>(matrix_data, <span class="st">"mydata.hdf5"</span>, <span class="st">"experiment/data"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="reading-hdf5-files" class="level4">
<h4 class="anchored" data-anchor-id="reading-hdf5-files">Reading HDF5 files</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect file structure</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">h5ls</span>(<span class="st">"mydata.hdf5"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Read entire dataset</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">h5read</span>(<span class="st">"mydata.hdf5"</span>, <span class="st">"experiment/data"</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Read subset (rows 1-100, all columns)</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>subset <span class="ot">&lt;-</span> <span class="fu">h5read</span>(<span class="st">"mydata.hdf5"</span>, <span class="st">"experiment/data"</span>, </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                 <span class="at">index =</span> <span class="fu">list</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="cn">NULL</span>))</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Close all open handles</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="simplified-operations-with-bigdatastatmeth" class="level3">
<h3 class="anchored" data-anchor-id="simplified-operations-with-bigdatastatmeth">Simplified Operations with BigDataStatMeth</h3>
<p>BigDataStatMeth provides high-level functions that abstract HDF5 complexity:</p>
<section id="create-hdf5-matrix-from-r-object" class="level4">
<h4 class="anchored" data-anchor-id="create-hdf5-matrix-from-r-object">Create HDF5 matrix from R object</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BigDataStatMeth)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create from matrix</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">10000</span>), <span class="dv">100</span>, <span class="dv">100</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">bdCreate_hdf5_matrix</span>(</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"analysis.hdf5"</span>,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> data,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">"input"</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">dataset =</span> <span class="st">"matrix_a"</span>,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">overwriteFile =</span> <span class="cn">TRUE</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="perform-operations-directly-on-hdf5" class="level4">
<h4 class="anchored" data-anchor-id="perform-operations-directly-on-hdf5">Perform operations directly on HDF5</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Matrix multiplication without loading into RAM</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">bdblockmult_hdf5</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"analysis.hdf5"</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">"input"</span>,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">A =</span> <span class="st">"matrix_a"</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">B =</span> <span class="st">"matrix_b"</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">blocksize =</span> <span class="dv">128</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Result automatically stored in HDF5</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Location: result$fn (filename), result$ds (dataset path)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="working-with-large-datasets" class="level4">
<h4 class="anchored" data-anchor-id="working-with-large-datasets">Working with large datasets</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example: PCA on large genomic data</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Data: 10,000 samples √ó 500,000 SNPs (too large for RAM)</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Create HDF5 from data</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">bdCreate_hdf5_matrix</span>(</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"genomics.hdf5"</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> snp_matrix,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">"genotypes"</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">dataset =</span> <span class="st">"all_snps"</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Perform PCA directly on HDF5</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>pca_result <span class="ot">&lt;-</span> <span class="fu">bdPCA_hdf5</span>(</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">filename =</span> <span class="st">"genomics.hdf5"</span>,</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> <span class="st">"genotypes"</span>,</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">dataset =</span> <span class="st">"all_snps"</span>,</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">k =</span> <span class="dv">20</span>,              <span class="co"># Number of components</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">bcenter =</span> <span class="cn">TRUE</span>,      <span class="co"># Center data</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">bscale =</span> <span class="cn">FALSE</span>,      <span class="co"># Don't scale</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">threads =</span> <span class="dv">4</span>          <span class="co"># Parallel processing</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Results stored in: /PCA/all_snps/</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="co"># - components</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a><span class="co"># - variance</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="co"># - eigenvalues</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
</section>
<section id="best-practices" class="level3">
<h3 class="anchored" data-anchor-id="best-practices">Best Practices</h3>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Warning</span>‚ö†Ô∏è Important Considerations
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Do:</strong> - ‚úÖ Always close HDF5 handles with <code>h5closeAll()</code> - ‚úÖ Use appropriate chunk sizes (10 KB - 1 MB) - ‚úÖ Apply compression for large datasets - ‚úÖ Organize data hierarchically (groups for different analyses) - ‚úÖ Use descriptive names for groups and datasets</p>
<p><strong>Don‚Äôt:</strong> - ‚ùå Keep files open longer than necessary - ‚ùå Use tiny chunks (&lt; 1 KB) - too much overhead - ‚ùå Use huge chunks (&gt; 10 MB) - defeats partial I/O - ‚ùå Mix data types in single analysis workflow - ‚ùå Forget to check available disk space</p>
</div>
</div>
</section>
<section id="common-pitfalls-and-solutions" class="level3">
<h3 class="anchored" data-anchor-id="common-pitfalls-and-solutions">Common Pitfalls and Solutions</h3>
<section id="problem-1-cannot-open-file" class="level4">
<h4 class="anchored" data-anchor-id="problem-1-cannot-open-file">Problem 1: ‚ÄúCannot open file‚Äù</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Problem</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">h5read</span>(<span class="st">"myfile.hdf5"</span>, <span class="st">"data"</span>)  <span class="co"># File locked</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Solution</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()  <span class="co"># Close all open handles</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">h5read</span>(<span class="st">"myfile.hdf5"</span>, <span class="st">"data"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="problem-2-out-of-memory" class="level4">
<h4 class="anchored" data-anchor-id="problem-2-out-of-memory">Problem 2: Out of memory</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Problem</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>big_data <span class="ot">&lt;-</span> <span class="fu">h5read</span>(<span class="st">"huge.hdf5"</span>, <span class="st">"data"</span>)  <span class="co"># 100 GB dataset</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Solution: Read in chunks with BigDataStatMeth</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">bdCrossprod_hdf5</span>(<span class="st">"huge.hdf5"</span>, <span class="st">"data"</span>, <span class="at">blocksize =</span> <span class="dv">1000</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</section>
<section id="problem-3-slow-performance" class="level4">
<h4 class="anchored" data-anchor-id="problem-3-slow-performance">Problem 3: Slow performance</h4>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Problem: Default chunking</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">h5createDataset</span>(<span class="st">"data.hdf5"</span>, <span class="st">"matrix"</span>, <span class="fu">c</span>(<span class="dv">10000</span>, <span class="dv">10000</span>))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Solution: Specify optimal chunks</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">h5createDataset</span>(<span class="st">"data.hdf5"</span>, <span class="st">"matrix"</span>, </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">dims =</span> <span class="fu">c</span>(<span class="dv">10000</span>, <span class="dv">10000</span>),</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">chunk =</span> <span class="fu">c</span>(<span class="dv">1000</span>, <span class="dv">1000</span>),  <span class="co"># 1000√ó1000 chunks</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">level =</span> <span class="dv">6</span>)               <span class="co"># Compression</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<hr>
</section>
</section>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>HDF5 is the <strong>optimal solution for big data analysis</strong> in R:</p>
<ul>
<li><strong>Scalable</strong>: Work with datasets larger than RAM</li>
<li><strong>Fast</strong>: Optimized I/O with chunking and compression<br>
</li>
<li><strong>Flexible</strong>: Hierarchical organization for complex workflows</li>
<li><strong>Integrated</strong>: BigDataStatMeth provides seamless statistical methods</li>
</ul>
<p><strong>Next steps:</strong></p>
<ul>
<li>Learn about <a href="../../fundamentals/hdf5/structure.html">HDF5 structure and organization</a></li>
<li>Explore <a href="../../fundamentals/hdf5/operations.html">HDF5 operations in detail</a></li>
<li>Optimize performance with <a href="../../fundamentals/hdf5/performance.html">chunking and compression</a></li>
</ul>
<hr>
</section>
<section id="additional-resources" class="level2">
<h2 class="anchored" data-anchor-id="additional-resources">Additional Resources</h2>
<section id="official-documentation" class="level3">
<h3 class="anchored" data-anchor-id="official-documentation">Official Documentation</h3>
<ul>
<li><a href="https://www.hdfgroup.org/">HDF5 Official Website</a></li>
<li><a href="https://bioconductor.org/packages/rhdf5/">rhdf5 Bioconductor</a></li>
<li><a href="https://portal.hdfgroup.org/display/HDF5/HDF5+User+Guides">HDF5 User Guide</a></li>
</ul>
</section>
<section id="tutorials" class="level3">
<h3 class="anchored" data-anchor-id="tutorials">Tutorials</h3>
<ul>
<li><a href="https://bioconductor.org/packages/release/bioc/vignettes/rhdf5/inst/doc/rhdf5.html">rhdf5 Tutorial</a></li>
<li><a href="https://docs.h5py.org/">HDF5 for Python Users</a></li>
</ul>
</section>
<section id="tools" class="level3">
<h3 class="anchored" data-anchor-id="tools">Tools</h3>
<ul>
<li><a href="https://www.hdfgroup.org/downloads/hdfview/">HDFView</a> - Visual HDF5 file browser</li>
<li><a href="https://portal.hdfgroup.org/display/HDF5/h5dump">h5dump</a> - Command-line inspection tool</li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "Óßã";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
            // default icon
            link.classList.add("external");
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>BigDataStatMeth - ISGlobal Bioinformatics Research Group</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Built with <a href="https://quarto.org/">Quarto</a></p>
</div>
  </div>
</footer>




<script src="../../site_libs/quarto-html/zenscroll-min.js"></script>
</body></html>