<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="BigDataStatMeth">
<title>Matrix operations, algebra and Statistical Methods for Big Data. Part II: working with HDF5 files • BigDataStatMeth</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Matrix operations, algebra and Statistical Methods for Big Data. Part II: working with HDF5 files">
<meta property="og:description" content="BigDataStatMeth">
<meta property="og:image" content="https://dpelegri.github.io/BigDataStatMeth/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">BigDataStatMeth</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.50</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BigDataStatMeth_hdf5.html">Matrix operations, algebra and Statistical Methods for Big Data. Part II: working with HDF5 files</a>
    <a class="dropdown-item" href="../articles/BigDataStatMeth_memory.html">Matrix operations, algebra and Statistical Methods for Big Data. Part I: working in memory</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="http://twitter.com/brgelab" aria-label="Twitter">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Matrix operations, algebra and Statistical Methods for Big Data. Part II: working with HDF5 files</h1>
                        <h4 data-toc-skip class="author">Dolors
Pelegri</h4>
            <address class="author_afil">
      Universitat Autonoma de Barcelona - Universitat Autonoma de
Barcelona (UAB)ISGlobal, Centre for Research in Environmental
Epidemiology (ISGlobal)Bioinformatics Research Group in Epidemiolgy
(BRGE)<br><h4 data-toc-skip class="author">Juan
R. Gonzalez</h4>
            <address class="author_afil">
      Universitat Autonoma de Barcelona - Universitat Autonoma de
Barcelona (UAB)ISGlobal, Centre for Research in Environmental
Epidemiology (ISGlobal)Bioinformatics Research Group in Epidemiolgy
(BRGE)<br><a class="author_email" href="mailto:#"></a><a href="mailto:juanr.gonzalez@isglobal.org" class="email">juanr.gonzalez@isglobal.org</a>
      </address>
                  
            <h4 data-toc-skip class="date">2022-10-25</h4>
      
      
      <div class="d-none name"><code>BigDataStatMeth_hdf5.Rmd</code></div>
    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Description of functions to perform matrix operations, algebra
      and basic statistical analyses using HDF5 data files.</p>
    </div>
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This package implements several matrix operations using
<code>Matrix</code> and <code>DelayMatrix</code> objects as well as HDF5
data files. Some basic algebra operations that can also be computed that
are useful to implement statistical analyses using standard
methodologies such as principal component analyses (PCA) or least
squares estimation. The package also contains specific statistical
methods mainly used in <code>omic</code> data analysis such as lasso
regression. All procedures to work with objects loaded in memory can be
found in <a href="https://github.com/isglobal-brge/BigDataStatMeth/blob/master/vignettes/BigDataStatMeth_memory.html" class="external-link">BigDataStatMeth_memory</a>
vignette.</p>
</div>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>The package requires other packages from <code>CRAN</code> and
<code>Bioconductor</code> to be installed.</p>
<ul>
<li>
<code>CRAN</code>: <code>Matrix</code>, <code>RcppEigen</code> and
<code>RSpectra</code>.</li>
<li>
<code>Bioconductor</code>: <code>HDF5array</code>,
<code>rhdf5</code>
</li>
</ul>
<p>The user can execute this code to install the required packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install BiocManager (if not previously installed)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Install required packages</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Matrix"</span>, <span class="st">"RcppEigen"</span>, <span class="st">"RSpectra"</span>,</span>
<span>                       <span class="st">"HDF5Array"</span>, <span class="st">"rhdf5"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Our package needs to be installed from source code. In such cases, a
collection of software (e.g. C, C++, Fortran, …) are required, mainly
for Windows users. These programs can be installed using <a href="https://cran.r-project.org/bin/windows/Rtools/" class="external-link">Rtools</a>.</p>
</div>
<div class="section level2">
<h2 id="install-package">Install package<a class="anchor" aria-label="anchor" href="#install-package"></a>
</h2>
<p>Once required packages and Rtools are installed,
<code>BigDataStatMeth</code> package can be installed from our GitHub
repository as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install devtools and load library (if not previously installed)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span> </span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install BigDataStatMeth </span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"isglobal-brge/BigDataStatMeth"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>First, let us start by loading the required packages to describe the
main capabilities of the package</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/rhdf5" class="external-link">rhdf5</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BigDataStatMeth</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="previous-knowledge">Previous knowledge<a class="anchor" aria-label="anchor" href="#previous-knowledge"></a>
</h2>
<div class="section level3">
<h3 id="hdf5-data-files">HDF5 data files<a class="anchor" aria-label="anchor" href="#hdf5-data-files"></a>
</h3>
<p>Hierarchical Data Format (HDF) is a set of file formats designed to
store and organize large amounts of data. It is supported by The HDF
Group, a non-profit corporation whose mission is to ensure continued
development of HDF5 technologies and the continued accessibility of data
stored in HDF.</p>
<p>HDF5 is a technology suite that makes possible the management of
extremely large and complex data collections, can accomodate virtually
every kind of data in a single file, sequences, images, SNP matrices,
and every other type of data and metadata associated with an
experiment.</p>
<p>There is no limit on the number or size of data objects in the
collection, giving great flexibility for omic data. Is high-performance
I/O with a rich set of integrated performance features that allow for
access time and storage space optimizations</p>
<p>HDF5 file structure include only <em>two major types of
object</em>:</p>
<ul>
<li><p><code>Datasets</code>, which are multidimensional arrays of a
homogeneous type. For example, datasets for omics data could be to
genomics, transcriptomics, epigenomics, proteomics and/or metabolomics
experiments</p></li>
<li><p><code>Groups</code>, which are container structures which can
hold datasets and other groups</p></li>
</ul>
<p>This results in a truly hierarchical, filesystem-like data format</p>
<div class="figure" style="text-align: center">
<img src="../../../../../../Mobile%20Documents/com~apple~CloudDocs/PROJECTES/Github/BigDataStatMeth/articles/fig/hdf5_squema.jpg" alt="HDF5 hierarchical structure" width="100%"><p class="caption">
HDF5 hierarchical structure
</p>
</div>
</div>
<div class="section level3">
<h3 id="basics-in-rhdf5">Basics in rhdf5<a class="anchor" aria-label="anchor" href="#basics-in-rhdf5"></a>
</h3>
<p><strong>Create hdf5 file</strong></p>
<p>We have implemented the <code><a href="../reference/bdCreate_hdf5_matrix_file.html">bdCreate_hdf5_matrix_file ()</a></code>
function to create an hdf5 file with a group and a dataset in a single
step. This function allows to create datasets from a standard R matrices
from <code>Matrix</code> objects.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rhdf5)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5234</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">600</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span>m,<span class="at">mean=</span><span class="dv">0</span>,<span class="at">sd=</span><span class="dv">1</span>), n,m)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co"># We also can create a dataset from R matrix object</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="fu">bdCreate_hdf5_matrix_file</span>(<span class="at">filename =</span> <span class="st">"robject.hdf5"</span>, </span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">object =</span> A, </span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>                          <span class="at">group =</span> <span class="st">"INPUT"</span>, </span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>                          <span class="at">dataset =</span> <span class="st">"A"</span>)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>File and dataset has been created</span></code></pre></div>
<p>We see 0 in the console, indicating that no errors where found when
creating the hdf5 file. Notice that a file called “robject.hdf5” will be
created in the working directory</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">list.files</span>()</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"BigDataStatMeth_hdf5.Rmd"</span>   <span class="st">"BigDataStatMeth_memory.Rmd"</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>] <span class="st">"colesterol_new.csv"</span>         <span class="st">"colesterol.csv"</span>            </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>] <span class="st">"fig"</span>                        <span class="st">"img_applications"</span>          </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">7</span>] <span class="st">"references_application.bib"</span> <span class="st">"references.bib"</span>            </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">9</span>] <span class="st">"robject.hdf5"</span>              </span></code></pre></div>
<p><strong>Add datasets in hdf5 file</strong></p>
<p>The function <code><a href="../reference/bdAdd_hdf5_matrix.html">bdAdd_hdf5_matrix ()</a></code> allows to add a
dataset in a existing hdf5 file. We can create the dataset in any group,
if group doesn’t exists in file, the group is created before append the
dataset.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5234</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span>m,<span class="at">mean=</span><span class="dv">0</span>,<span class="at">sd=</span><span class="dv">1</span>), n, m)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5234</span>)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">12000</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span>m,<span class="at">mean=</span><span class="dv">3</span>,<span class="at">sd=</span><span class="fl">0.5</span>), n, m)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="co"># We create another data file (delayed.hdf5) with a matrix A.</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># The group is called INPUT</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">bdCreate_hdf5_matrix_file</span>(<span class="at">filename =</span> <span class="st">"delayed.hdf5"</span>, </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                        <span class="at">object =</span> A, </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>                        <span class="at">group =</span> <span class="st">"INPUT"</span>, </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>                        <span class="at">dataset =</span> <span class="st">"A"</span>)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>File and dataset has been created</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="co"># And them, we add another matrix B to the same group</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="fu">bdAdd_hdf5_matrix</span>(<span class="at">object =</span> B, </span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>                <span class="at">filename =</span> <span class="st">"delayed.hdf5"</span>, </span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>                <span class="at">group =</span> <span class="st">"INPUT"</span>, </span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>                <span class="at">dataset =</span> <span class="st">"B"</span>)</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>Dataset has been added to the file</span></code></pre></div>
<p><strong>Create transposed matrix in hdf5 file</strong></p>
<p>The function <code>Add_hdf5_matrix ()</code> also allows to create a
dataset with a transposed matrix with the parameter
<code>transpose = TRUE</code>. For example, we can add a dataset in hdf5
with the B transposed matrix with :</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dataset BTransposed with B transposed matrix in delayed.hdf5 file at INPUT group</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">bdAdd_hdf5_matrix</span>(B, <span class="st">"delayed.hdf5"</span>, <span class="st">"INPUT"</span>, <span class="st">"BTransposed"</span>, <span class="at">transp =</span> <span class="cn">TRUE</span>);</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>Dataset has been added to the file</span></code></pre></div>
<p><strong>Open and get hdf5 content file</strong></p>
<p>We can open an existing file show contents and access data using
functions from <code>rhdf5</code> package. <code>rhdf5</code> is an R
interface for HDF5. The file must always be opened before working with
it.</p>
<p>With <code>htls ()</code> function we get the the list of an hdf5
content without the need of open the file.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Examine hierarchy before open file</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">h5ls</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>   group        name       otype dclass         dim</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span>      <span class="sc">/</span>       INPUT   H5I_GROUP                   </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">/</span>INPUT           A H5I_DATASET  FLOAT <span class="dv">100</span> x <span class="dv">10000</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="sc">/</span>INPUT           B H5I_DATASET  FLOAT  <span class="dv">50</span> x <span class="dv">12000</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> <span class="sc">/</span>INPUT BTransposed H5I_DATASET  FLOAT  <span class="dv">12000</span> x <span class="dv">50</span></span></code></pre></div>
<p>Once opened, it can be seen just typing its name. In that case, we
only get the current level in the hierarchical tree</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Open file</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>h5fdelay <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Show hdf5 hierarchy (groups)</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>h5fdelay</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>HDF5 FILE </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>        name <span class="sc">/</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>    filename </span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>   name     otype dclass dim</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span> INPUT H5I_GROUP           </span></code></pre></div>
<p>NOTE: We can also use <a href="https://www.hdfgroup.org/downloads/hdfview/" class="external-link">hdfview</a></p>
<p><strong>Access datasets data</strong></p>
<p>The $ operator can be used to access the next group level, this
operator reads the object from disk. We can assign the dataset contents
to an R object in order to work with it.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>Bdata <span class="ot">&lt;-</span> h5fdelay<span class="sc">$</span>INPUT<span class="sc">$</span>B</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>Bdata[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>         [,<span class="dv">1</span>]     [,<span class="dv">2</span>]     [,<span class="dv">3</span>]     [,<span class="dv">4</span>]     [,<span class="dv">5</span>]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">3.562285</span> <span class="fl">2.699083</span> <span class="fl">3.842990</span> <span class="fl">3.400242</span> <span class="fl">3.412452</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">2.702681</span> <span class="fl">2.313405</span> <span class="fl">2.747726</span> <span class="fl">3.044964</span> <span class="fl">3.035956</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="fl">3.057644</span> <span class="fl">3.251765</span> <span class="fl">3.260262</span> <span class="fl">3.322551</span> <span class="fl">3.739247</span></span></code></pre></div>
<p>Or we can get the data for the transposed version of this matrix
created before</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>BdataTrans <span class="ot">&lt;-</span> h5fdelay<span class="sc">$</span>INPUT<span class="sc">$</span>BTransposed</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>BdataTrans[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>         [,<span class="dv">1</span>]     [,<span class="dv">2</span>]     [,<span class="dv">3</span>]</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">3.562285</span> <span class="fl">2.702681</span> <span class="fl">3.057644</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">2.699083</span> <span class="fl">2.313405</span> <span class="fl">3.251765</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="fl">3.842990</span> <span class="fl">2.747726</span> <span class="fl">3.260262</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,] <span class="fl">3.400242</span> <span class="fl">3.044964</span> <span class="fl">3.322551</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,] <span class="fl">3.412452</span> <span class="fl">3.035956</span> <span class="fl">3.739247</span></span></code></pre></div>
<blockquote>
<p>IMPORTANT NOTE: This is a very interesting feature. Imaging you wan
to compute the mean of each column in the matrix A. If you want to do
this in R, you first need to load the file and then use any of the
existing functions in R. If A is big, loading all the matrix in R can be
inneficient since you actually only need the information in one
column</p>
</blockquote>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="cf">function</span>(i){</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(h5fdelay<span class="sc">$</span>INPUT<span class="sc">$</span>A[,i])</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(out)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="fl">0.05107500</span>  <span class="fl">0.25916486</span>  <span class="fl">0.14013419</span>  <span class="fl">0.09320278</span> <span class="sc">-</span><span class="fl">0.13489573</span>  <span class="fl">0.10595091</span></span></code></pre></div>
<p>There are efficient ways of reading a subset of data without the need
of reading the entire file. To this end, see this <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/rhdf5/inst/doc/practical_tips.html" class="external-link">Practical
tips</a>.</p>
<p><strong>Close hdf5 file</strong></p>
<p>After work with hdf5 <strong>the user always must close the
file</strong>. We can close only one file with <code>H5Fclose</code>
function or close all hdf5 opened files with <code>h5closeAll</code>
function</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Close delayed.hdf5 file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fclose.html" class="external-link">H5Fclose</a></span><span class="op">(</span><span class="va">h5fdelay</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Open 2 files and close all</span></span>
<span><span class="va">h5fdelay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fopen.html" class="external-link">H5Fopen</a></span><span class="op">(</span><span class="st">"delayed.hdf5"</span><span class="op">)</span></span>
<span><span class="va">h5fr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fopen.html" class="external-link">H5Fopen</a></span><span class="op">(</span><span class="st">"robject.hdf5"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5closeAll.html" class="external-link">h5closeAll</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>More information about working with hdf5 files in R can be found in
<a href="https://www.bioconductor.org/packages/release/bioc/html/rhdf5.html" class="external-link">several
vignettes</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="convert-text-file-to-hdf5">Convert text file to HDF5<a class="anchor" aria-label="anchor" href="#convert-text-file-to-hdf5"></a>
</h2>
<p>Usually we have files with a large amount of data to be loaded into
R. For Big data problem, R can’t deal with this data so we cannot
efficiently work with then. In order to resolve this issue, you can work
directly from data on disk with datasets stored in an HDF5 data
file.</p>
<p>You can easily change your text files to HDF5 using the function
<code><a href="../reference/bdImport_text_to_hdf5.html">bdImport_text_to_hdf5 ()</a></code>, to do that, we only have to
define the input file with the <code>filename</code> parameter and the
HDF5 destination dataset with parameters <code>outputfile</code> for the
output file name, <code>outGroup</code> with the group to store the
dataset and the dataset name with the parameter <code>outDataset</code>.
It should be considered that:</p>
<ul>
<li><p>This function <strong>only allows to import numeric data except
for rownames and columnames where character are
allowed</strong>.</p></li>
<li><p>Data can be delimited by different characters. By default, we use
tabs (“\t”) but it can be changed easily with <code>sep</code>
parameter.</p></li>
</ul>
<p>In this example we convert the “colesterol.csv” file and we will save
the data to a file ‘colesterol_file.hdf5’ under the group
‘COLESTEROLGROUP’ in the dataset ‘COLESTEROLDATASET’. In this example
the text file contains column names so we set the parameter
<code>header = TRUE</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bdImport_text_to_hdf5</span>(<span class="at">filename =</span> <span class="st">"colesterol.csv"</span>, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sep=</span><span class="st">','</span>, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">outputfile =</span> <span class="st">"colesterol_file.hdf5"</span>, </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">outGroup =</span> <span class="st">"COLESTEROL"</span>, </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">outDataset =</span> <span class="st">"COLESTEROLDATA"</span>, </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>COLESTEROLDATA dataset has been imported successfully.</span></code></pre></div>
<p>If we observe the new file <code>colesterol_file.hdf5</code> and its
content</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">h5ls</span>(<span class="st">"colesterol_file.hdf5"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>                                 group                     name       otype</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span>                                    <span class="sc">/</span>               COLESTEROL   H5I_GROUP</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>                          <span class="sc">/</span>COLESTEROL .COLESTEROLDATA_dimnames   H5I_GROUP</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="sc">/</span>COLESTEROL<span class="sc">/</span>.COLESTEROLDATA_dimnames                        <span class="dv">2</span> H5I_DATASET</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>                          <span class="sc">/</span>COLESTEROL           COLESTEROLDATA H5I_DATASET</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>    dclass       dim</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span>                   </span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>                   </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> COMPOUND        <span class="dv">10</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>    FLOAT <span class="dv">1000</span> x <span class="dv">10</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a><span class="co"># We can open the file and have access to the data</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>h5Colesterol <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(<span class="st">"colesterol_file.hdf5"</span>)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Show hdf5 content dataset</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>h5Colesterol<span class="sc">$</span>COLESTEROL<span class="sc">$</span>COLESTEROLDATA[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>         [,<span class="dv">1</span>]     [,<span class="dv">2</span>]     [,<span class="dv">3</span>]      [,<span class="dv">4</span>]     [,<span class="dv">5</span>]     [,<span class="dv">6</span>]</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">223.7348</span> <span class="fl">55.25039</span> <span class="fl">14.90246</span> <span class="fl">0.9026095</span> <span class="fl">10.22067</span> <span class="fl">1.450970</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">248.1820</span> <span class="fl">53.47404</span> <span class="fl">25.12592</span> <span class="fl">0.8710345</span> <span class="fl">17.10225</span> <span class="fl">1.002655</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="fl">180.2071</span> <span class="fl">58.54268</span> <span class="fl">12.95197</span> <span class="fl">0.8882435</span> <span class="fl">11.21530</span> <span class="fl">1.073924</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,] <span class="fl">200.1522</span> <span class="fl">62.58284</span> <span class="fl">28.27819</span> <span class="fl">0.9361357</span> <span class="fl">12.79399</span> <span class="fl">1.129373</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,] <span class="fl">234.3281</span> <span class="fl">68.70254</span> <span class="fl">13.21404</span> <span class="fl">0.7775238</span> <span class="fl">11.39689</span> <span class="fl">1.235655</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Show colnames </span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(h5Colesterol<span class="sc">$</span>COLESTEROL<span class="sc">$</span>.COLESTEROLDATA_dimnames<span class="sc">$</span><span class="st">`</span><span class="at">2</span><span class="st">`</span>)</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>             chr</span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="st">"TCholesterol"</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>          <span class="st">"Age"</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>      <span class="st">"Insulin"</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>   <span class="st">"Creatinine"</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>          <span class="st">"BUN"</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>         <span class="st">"LLDR"</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a><span class="fu">H5Fclose</span>(h5Colesterol)</span></code></pre></div>
<p>NOTE: We can overwrite an existing hdf5 file with new data by setting
<code>overwrite = TRUE</code></p>
</div>
<div class="section level2">
<h2 id="operation-with-hdf5-files">Operation with HDF5 files<a class="anchor" aria-label="anchor" href="#operation-with-hdf5-files"></a>
</h2>
<p>Once our data are available as hdf5 files, we can operate directly on
it without having data on memory (i.e we have access from disk. In
<code>BigDataStatMeth</code> we have implemented most of the common
matrix operations and algebra functions to help users how are not
familiar with this type of implementation. Next section provides several
examples</p>
<div class="section level3">
<h3 id="hdf5-matrix-multiplication">HDF5 Matrix Multiplication<a class="anchor" aria-label="anchor" href="#hdf5-matrix-multiplication"></a>
</h3>
<p>In this section, different products of matrices and vectors are
introduced. The methods implement different strategies including block
multiplication algorithms and the use of parallel implementations with
hdf5 files datasets.</p>
<p><code>BigDataStatMeth</code> allows to make an hdf5 matrix
multiplication with :</p>
<ul>
<li>two on memory matrices, previously saving the data in a file</li>
<li>two matrices inside an hdf5 file.</li>
</ul>
<div class="section level4">
<h4 id="hdf5-matrix-multiplication-from-data-in-memory">Hdf5 Matrix multiplication from data in memory<a class="anchor" aria-label="anchor" href="#hdf5-matrix-multiplication-from-data-in-memory"></a>
</h4>
<p>To work with big matrices <code><a href="../reference/bdblockmult.html">bdblockmult()</a></code> saves matrices
in hdf5 file format in order to be able to operate with them in blocks
and not overload the memory, by default are considered large matrices if
the number of rows or columns is greater than 5000, but it can be
changed with <code>bigmatrix</code> argument.</p>
<p>In that case, by default the function <code><a href="../reference/bdblockmult.html">bdblockmult()</a></code>, by
default, this function creates the <em>tmp_blockmult.hdf5</em> file with
two groups the <em>INPUT</em> group with datasets <em>A</em> and
<em>B</em> with original matrices and the <em>OUTPUT</em> group with
dataset <em>C</em> with the results. The file name can be set with
<code>outputfile</code> parameter.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fl">10000</span></span>
<span><span class="va">Abig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">n</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">n</span>, ncol<span class="op">=</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">Bbig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">p</span><span class="op">)</span>, nrow<span class="op">=</span><span class="va">n</span>, ncol<span class="op">=</span><span class="va">p</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We want to force it to run in memory</span></span>
<span><span class="va">AxBNOBig</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdblockmult.html">bdblockmult</a></span><span class="op">(</span><span class="va">Abig</span>, <span class="va">Bbig</span>, onmemory <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># We want to run it on disk</span></span>
<span><span class="va">AxBBig3000</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdblockmult.html">bdblockmult</a></span><span class="op">(</span><span class="va">Abig</span>, <span class="va">Bbig</span>, onmemory <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>Depending on whether the calculation has been performed directly in
memory or from an hdf5 file, the returned object type is different.</p>
<p>If we work in memory results are returned as a current r matrix
object,</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(AxBNOBig)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"matrix"</span> <span class="st">"array"</span> </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>AxBNOBig[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>           [,<span class="dv">1</span>]       [,<span class="dv">2</span>]        [,<span class="dv">3</span>]      [,<span class="dv">4</span>]       [,<span class="dv">5</span>]</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]  <span class="fl">32.312471</span> <span class="sc">-</span><span class="fl">22.406542</span>  <span class="sc">-</span><span class="fl">7.4709272</span> <span class="sc">-</span><span class="fl">19.58841</span>  <span class="fl">19.578910</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="sc">-</span><span class="fl">20.569618</span>  <span class="fl">28.035882</span>   <span class="fl">0.3045435</span>  <span class="fl">73.34118</span> <span class="sc">-</span><span class="fl">30.459568</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="sc">-</span><span class="fl">10.311862</span>   <span class="fl">5.793771</span> <span class="sc">-</span><span class="fl">37.5088071</span>  <span class="fl">31.89941</span>   <span class="fl">4.087295</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]  <span class="fl">52.214687</span> <span class="sc">-</span><span class="fl">23.867445</span> <span class="sc">-</span><span class="fl">34.2711649</span> <span class="sc">-</span><span class="fl">36.50421</span>  <span class="fl">46.159705</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]  <span class="sc">-</span><span class="fl">7.148278</span> <span class="sc">-</span><span class="fl">26.950741</span>  <span class="fl">18.8095259</span> <span class="sc">-</span><span class="fl">22.66078</span> <span class="sc">-</span><span class="fl">10.875507</span></span></code></pre></div>
<p>if we work in disk, we return a list with the filnme and the dataset
that contains the results inside the hdf5 file. Then we can open the
hdf5 file and get the result matrix and the original matrices,</p>
<p>To work in R with hdf5 data object we can use the <code>rhdf5</code>
function library from <code>bioconductor</code>, we can open file and
read all the content</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>h5data <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(AxBBig3000<span class="sc">$</span>file)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co"># We can get where data is stored</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>AxBBig3000<span class="sc">$</span>dataset</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"OUTPUT//C"</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co"># We observe that the data is in folder OUTPUT dataset C</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>reshdf5 <span class="ot">&lt;-</span> h5data<span class="sc">$</span>OUTPUT<span class="sc">$</span>C</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>reshdf5[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>           [,<span class="dv">1</span>]       [,<span class="dv">2</span>]        [,<span class="dv">3</span>]      [,<span class="dv">4</span>]       [,<span class="dv">5</span>]</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]  <span class="fl">32.312471</span> <span class="sc">-</span><span class="fl">22.406542</span>  <span class="sc">-</span><span class="fl">7.4709272</span> <span class="sc">-</span><span class="fl">19.58841</span>  <span class="fl">19.578910</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="sc">-</span><span class="fl">20.569618</span>  <span class="fl">28.035882</span>   <span class="fl">0.3045435</span>  <span class="fl">73.34118</span> <span class="sc">-</span><span class="fl">30.459568</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="sc">-</span><span class="fl">10.311862</span>   <span class="fl">5.793771</span> <span class="sc">-</span><span class="fl">37.5088071</span>  <span class="fl">31.89941</span>   <span class="fl">4.087295</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]  <span class="fl">52.214687</span> <span class="sc">-</span><span class="fl">23.867445</span> <span class="sc">-</span><span class="fl">34.2711649</span> <span class="sc">-</span><span class="fl">36.50421</span>  <span class="fl">46.159705</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]  <span class="sc">-</span><span class="fl">7.148278</span> <span class="sc">-</span><span class="fl">26.950741</span>  <span class="fl">18.8095259</span> <span class="sc">-</span><span class="fl">22.66078</span> <span class="sc">-</span><span class="fl">10.875507</span></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(reshdf5, AxBNOBig)</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
<p>Remember that it is <strong>It is important to close all dataset,
group, and file handles when not used</strong> anymore,</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Close file</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fclose.html" class="external-link">H5Fclose</a></span><span class="op">(</span><span class="va">h5data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="hdf5-matrix-multiplication-from-data-stored-in-a-file">Hdf5 Matrix multiplication from data stored in a file<a class="anchor" aria-label="anchor" href="#hdf5-matrix-multiplication-from-data-stored-in-a-file"></a>
</h4>
<p>If we have the matrices stored in hdf5 data file, we can use the
function <code>bdblockmult_hdf5</code>. This function allows to perform
a matrix multiplication from two matrices stored in a file.</p>
<p>To use the <code>bdblockmult_hdf5</code>, we have to set the
parameters <code>filename</code> with the file that contains the
matrices, <code>group</code>, here, we have to indicate under which
group are stored the matrices and the datasets names where data is
stored parameters <code>a</code> and <code>b</code>. By default, the
results group can be set by <code>outgroup</code> parameter by default,
the output group is <em>OUTPUT</em> .</p>
<p>To show how this function works, we will use the matrices created
with <code>bdblockmult</code> function previously used and we will store
the results under the ‘HDF5_RES’ group</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">bdblockmult_hdf5</span>(<span class="at">filename =</span> <span class="st">'tmp_blockmult.hdf5'</span>, <span class="at">group=</span><span class="st">"INPUT"</span>,</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">a=</span><span class="st">"A"</span>, <span class="at">b=</span><span class="st">"B"</span>, <span class="at">outgroup =</span> <span class="st">'HDF5_RES'</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We show the hdf5 content</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a> <span class="fu">h5ls</span>(res<span class="sc">$</span>file)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>      group     name       otype dclass         dim</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="dv">0</span>         <span class="sc">/</span> HDF5_RES   H5I_GROUP                   </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">/</span>HDF5_RES    A_x_B H5I_DATASET  FLOAT <span class="dv">500</span> x <span class="dv">10000</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>         <span class="sc">/</span>    INPUT   H5I_GROUP                   </span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>    <span class="sc">/</span>INPUT        A H5I_DATASET  FLOAT   <span class="dv">500</span> x <span class="dv">500</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>    <span class="sc">/</span>INPUT        B H5I_DATASET  FLOAT <span class="dv">500</span> x <span class="dv">10000</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>         <span class="sc">/</span>   OUTPUT   H5I_GROUP                   </span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>   <span class="sc">/</span>OUTPUT        C H5I_DATASET  FLOAT <span class="dv">500</span> x <span class="dv">10000</span></span></code></pre></div>
<p>Now, we compare the obtained results with
<code>bdblockmult_hdf5</code> and <code>bdblockmult</code> functions
.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get content</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>h5data <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(res<span class="sc">$</span>file)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># We can get where data is stored</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>dataset</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"HDF5_RES/A_x_B"</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a><span class="co"># We get the results with bdblockmult (previous step)</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>resblockmult <span class="ot">&lt;-</span> h5data<span class="sc">$</span>OUTPUT<span class="sc">$</span>C</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># and the results obtained from bdblockmult_hdf5</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>resblockmult_hdf5 <span class="ot">&lt;-</span> h5data<span class="sc">$</span>HDF5_RES<span class="sc">$</span>A_x_B</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>resblockmult_hdf5[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>           [,<span class="dv">1</span>]       [,<span class="dv">2</span>]        [,<span class="dv">3</span>]      [,<span class="dv">4</span>]       [,<span class="dv">5</span>]</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]  <span class="fl">32.312471</span> <span class="sc">-</span><span class="fl">22.406542</span>  <span class="sc">-</span><span class="fl">7.4709272</span> <span class="sc">-</span><span class="fl">19.58841</span>  <span class="fl">19.578910</span></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="sc">-</span><span class="fl">20.569618</span>  <span class="fl">28.035882</span>   <span class="fl">0.3045435</span>  <span class="fl">73.34118</span> <span class="sc">-</span><span class="fl">30.459568</span></span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="sc">-</span><span class="fl">10.311862</span>   <span class="fl">5.793771</span> <span class="sc">-</span><span class="fl">37.5088071</span>  <span class="fl">31.89941</span>   <span class="fl">4.087295</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]  <span class="fl">52.214687</span> <span class="sc">-</span><span class="fl">23.867445</span> <span class="sc">-</span><span class="fl">34.2711649</span> <span class="sc">-</span><span class="fl">36.50421</span>  <span class="fl">46.159705</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]  <span class="sc">-</span><span class="fl">7.148278</span> <span class="sc">-</span><span class="fl">26.950741</span>  <span class="fl">18.8095259</span> <span class="sc">-</span><span class="fl">22.66078</span> <span class="sc">-</span><span class="fl">10.875507</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the file</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(resblockmult, resblockmult_hdf5)</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
<p>We obtain the same result using <code>bdblockmult</code> and
<code>bdblockmult_hdf5</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="cross-product-and-transposed-cross-product-with-hdf5-files">Cross-product and Transposed Cross-product with hdf5 files<a class="anchor" aria-label="anchor" href="#cross-product-and-transposed-cross-product-with-hdf5-files"></a>
</h3>
<p>To perform a cross-product <span class="math inline">\(C = A^t
A\)</span> you can call <code><a href="../reference/bdCrossprod_hdf5.html">bdCrossprod_hdf5()</a></code> function. We
just have to set the parameters <code>filename</code>,
<code>group</code> and <code>A</code> with the file, group and the
dataset that we want to use and optionally the <code>outgroup</code>
parameter to set the output group where we want to store results.</p>
<p>The result dataset is stored with
<em>CrossProd_<matrixname>x<matrixname></matrixname></matrixname></em> pattern.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdCrossprod_hdf5.html">bdCrossprod_hdf5</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">'robject.hdf5'</span>, </span>
<span>                        group<span class="op">=</span><span class="st">"INPUT"</span>, </span>
<span>                        A<span class="op">=</span><span class="st">"A"</span>, </span>
<span>                        outgroup <span class="op">=</span> <span class="st">"RESULTS"</span><span class="op">)</span></span></code></pre></div>
<p>We obtain the expected values computed using <code>crossprod</code>
function.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Get content</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>h5data <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(res<span class="sc">$</span>file)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We can get wher data is stored</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>dataset</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"RESULTS//CrossProd_A"</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co"># We get the Crossprod Results and the original matrix</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>resCProd <span class="ot">&lt;-</span> h5data<span class="sc">$</span>RESULTS<span class="sc">$</span>CrossProd_AxA</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> h5data<span class="sc">$</span>INPUT<span class="sc">$</span>A</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the file</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Show results</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>resCProd[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>            [,<span class="dv">1</span>]       [,<span class="dv">2</span>]        [,<span class="dv">3</span>]      [,<span class="dv">4</span>]       [,<span class="dv">5</span>]</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">505.6616915</span>  <span class="fl">10.611020</span>  <span class="sc">-</span><span class="fl">0.3666995</span>  <span class="fl">13.24150</span>   <span class="fl">8.001664</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]  <span class="fl">10.6110197</span> <span class="fl">489.229418</span>  <span class="sc">-</span><span class="fl">9.4790008</span> <span class="sc">-</span><span class="fl">10.69272</span> <span class="sc">-</span><span class="fl">22.537412</span></span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,]  <span class="sc">-</span><span class="fl">0.3666995</span>  <span class="sc">-</span><span class="fl">9.479001</span> <span class="fl">497.0000736</span> <span class="sc">-</span><span class="fl">16.47190</span>  <span class="sc">-</span><span class="fl">1.484056</span></span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]  <span class="fl">13.2415011</span> <span class="sc">-</span><span class="fl">10.692717</span> <span class="sc">-</span><span class="fl">16.4719016</span> <span class="fl">583.94688</span>  <span class="sc">-</span><span class="fl">4.379890</span></span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]   <span class="fl">8.0016642</span> <span class="sc">-</span><span class="fl">22.537412</span>  <span class="sc">-</span><span class="fl">1.4840559</span>  <span class="sc">-</span><span class="fl">4.37989</span> <span class="fl">525.367876</span></span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(resCProd, <span class="fu">crossprod</span>(A))</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
<p>You may also compute the Crossprod with two different matrix <span class="math inline">\(C = A^t B\)</span> ,</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5234</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">600</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span>m,<span class="at">mean=</span><span class="dv">0</span>,<span class="at">sd=</span><span class="dv">1</span>), n,m)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="fu">bdAdd_hdf5_matrix</span>(B, <span class="at">filename =</span> <span class="st">'robject.hdf5'</span>, <span class="at">group=</span><span class="st">"INPUT"</span>, <span class="at">dataset =</span> <span class="st">"B2"</span>)</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>Dataset has been added to the file</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Get Crossprod with two matrix</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">bdCrossprod_hdf5</span>(<span class="at">filename =</span> <span class="st">'robject.hdf5'</span>, <span class="at">group=</span><span class="st">"INPUT"</span>, <span class="at">A=</span><span class="st">"A"</span>, </span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>                        <span class="at">groupB =</span> <span class="st">"INPUT"</span>, <span class="at">B =</span> <span class="st">"B2"</span>, <span class="at">outgroup =</span> <span class="st">"RESULTS"</span>)</span></code></pre></div>
<p>We obtain the expected values computed using <code>crossprod</code>
function.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Get content</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>h5data <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(res<span class="sc">$</span>file)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We can get wher data is stored</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>dataset</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"RESULTS//CrossProd_AxB2"</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="co"># We get the Crossprod Results and the original matrix</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>resCProd <span class="ot">&lt;-</span> h5data<span class="sc">$</span>RESULTS<span class="sc">$</span>CrossProd_AxB2</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> h5data<span class="sc">$</span>INPUT<span class="sc">$</span>A</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>B <span class="ot">&lt;-</span> h5data<span class="sc">$</span>INPUT<span class="sc">$</span>B2</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the file</span></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Show results</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>resCProd[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>            [,<span class="dv">1</span>]       [,<span class="dv">2</span>]        [,<span class="dv">3</span>]      [,<span class="dv">4</span>]       [,<span class="dv">5</span>]</span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">505.6616915</span>  <span class="fl">10.611020</span>  <span class="sc">-</span><span class="fl">0.3666995</span>  <span class="fl">13.24150</span>   <span class="fl">8.001664</span></span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]  <span class="fl">10.6110197</span> <span class="fl">489.229418</span>  <span class="sc">-</span><span class="fl">9.4790008</span> <span class="sc">-</span><span class="fl">10.69272</span> <span class="sc">-</span><span class="fl">22.537412</span></span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,]  <span class="sc">-</span><span class="fl">0.3666995</span>  <span class="sc">-</span><span class="fl">9.479001</span> <span class="fl">497.0000736</span> <span class="sc">-</span><span class="fl">16.47190</span>  <span class="sc">-</span><span class="fl">1.484056</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]  <span class="fl">13.2415011</span> <span class="sc">-</span><span class="fl">10.692717</span> <span class="sc">-</span><span class="fl">16.4719016</span> <span class="fl">583.94688</span>  <span class="sc">-</span><span class="fl">4.379890</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]   <span class="fl">8.0016642</span> <span class="sc">-</span><span class="fl">22.537412</span>  <span class="sc">-</span><span class="fl">1.4840559</span>  <span class="sc">-</span><span class="fl">4.37989</span> <span class="fl">525.367876</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(resCProd, <span class="fu">crossprod</span>(A,B))</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
<p>you may also set use <code><a href="../reference/bdtCrossprod_hdf5.html">bdtCrossprod_hdf5()</a></code> function to get
a transposed cross-product <span class="math inline">\(C = A
A^t\)</span> . In that case the result dataset is stored with
<em>tCrossProd_<matrixname>x<matrixname></matrixname></matrixname></em> pattern.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdtCrossprod_hdf5.html">bdtCrossprod_hdf5</a></span><span class="op">(</span>filename <span class="op">=</span> <span class="st">'robject.hdf5'</span>, group<span class="op">=</span><span class="st">"INPUT"</span>, A<span class="op">=</span><span class="st">"A"</span>, outgroup <span class="op">=</span> <span class="st">"RESULTS"</span><span class="op">)</span></span></code></pre></div>
<p>We obtain the expected values computed using <code>crossprod</code>
function.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Get content</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>h5data <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(res<span class="sc">$</span>file)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We can get wher data is stored</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>dataset</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"RESULTS//tCrossProd_A"</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co"># We get the Crossprod Results and the original matrix</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>restCProd <span class="ot">&lt;-</span> h5data<span class="sc">$</span>RESULTS<span class="sc">$</span>tCrossProd_AxA</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> h5data<span class="sc">$</span>INPUT<span class="sc">$</span>A</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the file</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Show results</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>restCProd[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]      [,<span class="dv">3</span>]      [,<span class="dv">4</span>]      [,<span class="dv">5</span>]</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">578.58545</span>  <span class="fl">17.23237</span> <span class="sc">-</span><span class="fl">13.95637</span> <span class="sc">-</span><span class="fl">27.63535</span>  <span class="fl">56.70321</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]  <span class="fl">17.23237</span> <span class="fl">603.99446</span> <span class="sc">-</span><span class="fl">43.22671</span>  <span class="fl">29.50159</span> <span class="sc">-</span><span class="fl">38.76707</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="sc">-</span><span class="fl">13.95637</span> <span class="sc">-</span><span class="fl">43.22671</span> <span class="fl">558.64887</span>  <span class="fl">19.51781</span>  <span class="fl">21.23500</span></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,] <span class="sc">-</span><span class="fl">27.63535</span>  <span class="fl">29.50159</span>  <span class="fl">19.51781</span> <span class="fl">645.86835</span> <span class="sc">-</span><span class="fl">40.12507</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]  <span class="fl">56.70321</span> <span class="sc">-</span><span class="fl">38.76707</span>  <span class="fl">21.23500</span> <span class="sc">-</span><span class="fl">40.12507</span> <span class="fl">542.25119</span></span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(restCProd, <span class="fu">tcrossprod</span>(A))</span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
<p>You may also compute the tCrossprod with two different matrix <span class="math inline">\(C = A B^t\)</span></p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5234</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">600</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>B <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span>m,<span class="at">mean=</span><span class="dv">0</span>,<span class="at">sd=</span><span class="dv">1</span>), n,m)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="fu">bdAdd_hdf5_matrix</span>(B, <span class="at">filename =</span> <span class="st">'robject.hdf5'</span>, <span class="at">group=</span><span class="st">"INPUT"</span>, <span class="at">dataset =</span> <span class="st">"B3"</span>)</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>Dataset has been added to the file</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Get Crossprod with two matrix</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">bdtCrossprod_hdf5</span>(<span class="at">filename =</span> <span class="st">'robject.hdf5'</span>, <span class="at">group=</span><span class="st">"INPUT"</span>, <span class="at">A=</span><span class="st">"A"</span>, </span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>                         <span class="at">groupB =</span> <span class="st">"INPUT"</span>, <span class="at">B =</span> <span class="st">"B3"</span>, <span class="at">outgroup =</span> <span class="st">"RESULTS"</span>)</span></code></pre></div>
<p>We obtain the expected values computed using <code>crossprod</code>
function.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Get content</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>h5data <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(res<span class="sc">$</span>file)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="co"># We can get wher data is stored</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>res<span class="sc">$</span>dataset</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">"RESULTS//tCrossProd_AxB3"</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a><span class="co"># We get the Crossprod Results and the original matrix</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>restCProd <span class="ot">&lt;-</span> h5data<span class="sc">$</span>RESULTS<span class="sc">$</span>tCrossProd_AxB3</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>A <span class="ot">&lt;-</span> h5data<span class="sc">$</span>INPUT<span class="sc">$</span>A</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>B <span class="ot">&lt;-</span> h5data<span class="sc">$</span>INPUT<span class="sc">$</span>B3</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Close the file</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Show results</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a>restCProd[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]      [,<span class="dv">3</span>]      [,<span class="dv">4</span>]      [,<span class="dv">5</span>]</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,] <span class="fl">578.58545</span>  <span class="fl">17.23237</span> <span class="sc">-</span><span class="fl">13.95637</span> <span class="sc">-</span><span class="fl">27.63535</span>  <span class="fl">56.70321</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]  <span class="fl">17.23237</span> <span class="fl">603.99446</span> <span class="sc">-</span><span class="fl">43.22671</span>  <span class="fl">29.50159</span> <span class="sc">-</span><span class="fl">38.76707</span></span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,] <span class="sc">-</span><span class="fl">13.95637</span> <span class="sc">-</span><span class="fl">43.22671</span> <span class="fl">558.64887</span>  <span class="fl">19.51781</span>  <span class="fl">21.23500</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,] <span class="sc">-</span><span class="fl">27.63535</span>  <span class="fl">29.50159</span>  <span class="fl">19.51781</span> <span class="fl">645.86835</span> <span class="sc">-</span><span class="fl">40.12507</span></span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]  <span class="fl">56.70321</span> <span class="sc">-</span><span class="fl">38.76707</span>  <span class="fl">21.23500</span> <span class="sc">-</span><span class="fl">40.12507</span> <span class="fl">542.25119</span></span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(restCProd, <span class="fu">tcrossprod</span>(A,B))</span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="singular-value-decomposition-svd">Singular Value Decomposition (SVD)<a class="anchor" aria-label="anchor" href="#singular-value-decomposition-svd"></a>
</h3>
<p>The SVD of an <span class="math inline">\(m \times n\)</span> real or
complex matrix <span class="math inline">\(A\)</span> is a factorization
of the form:</p>
<p><span class="math display">\[U\Sigma { V }^{ T }\]</span></p>
<p>where :</p>
<ul>
<li>
<span class="math inline">\(U\)</span> is a <span class="math inline">\(m \times m\)</span> real or complex unitary
matrix</li>
<li>
<span class="math inline">\(\Sigma\)</span> is a <span class="math inline">\(m \times n\)</span> rectangular diagonal matrix
with non-negative real numbers on the diagonal</li>
<li>
<span class="math inline">\(V\)</span> is a <span class="math inline">\(n \times n\)</span> real or complex unitary
matrix.</li>
</ul>
<p>Notice that:</p>
<ul>
<li>The diagonal entries <span class="math inline">\(\sigma_i\)</span>
of <span class="math inline">\(\Sigma\)</span> are known as the singular
values of <span class="math inline">\(A\)</span>.</li>
<li>The columns of <span class="math inline">\(U\)</span> are called the
left-singular vectors of <span class="math inline">\(A\)</span>.</li>
<li>The columns of <span class="math inline">\(V\)</span> are called the
right-singular vectors of <span class="math inline">\(A\)</span>.</li>
</ul>
<div class="section level4">
<h4 id="block-singular-values-decomposition">Block Singular Values Decomposition<a class="anchor" aria-label="anchor" href="#block-singular-values-decomposition"></a>
</h4>
<p>This method was developed by M. A. Iwen and B. W. Ong (2016). The
authors introduced a distributed and incremental SVD algorithm that is
useful for agglomerative data analysis on large networks. The algorithm
calculates the singular values and left singular vectors of a matrix A,
by first, partitioning it by columns. This creates a set of submatrices
of A with the same number of rows, but only some of its columns. After
that, the SVD of each of the submatrices is computed. The final step
consists of combining the results obtained by merging them again and
computing the SVD of the resulting matrix.</p>
<div class="figure" style="text-align: center">
<img src="../../../../../../Mobile%20Documents/com~apple~CloudDocs/PROJECTES/Github/BigDataStatMeth/articles/fig/blocksvd.png" alt="Flowchart for a two-level hierarchical Block SVD algorithm" width="100%"><p class="caption">
Flowchart for a two-level hierarchical Block SVD algorithm
</p>
</div>
<p>This method is implemented in <code>bdSVD_hdf5</code> function, this
function works directly on hdf5 data format, loading in memory only the
data to perform calculations and saving the results again in the hdf5
file for later use.</p>
<p>We have to indicate the file to work with, the dataset name and the
group where the dataset is located :</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dataframe data with 'odata' matrix in delayed hdf5 file at OMIC group</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">5234</span>)</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">15000</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>odata <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n<span class="sc">*</span>m, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span><span class="dv">1</span>), n,m)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="fu">bdAdd_hdf5_matrix</span>(odata, <span class="st">"delayed.hdf5"</span>, <span class="st">"OMICS"</span>, <span class="st">"data"</span>)</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>Dataset has been added to the file</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Direct from hdf5 data file</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>svdh5 <span class="ot">&lt;-</span> <span class="fu">bdSVD_hdf5</span>(<span class="st">"delayed.hdf5"</span>, <span class="st">"OMICS"</span>, <span class="st">"data"</span>)</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a><span class="co"># with R implementation from data in memory</span></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="co"># get results svd (d)</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>svd_hdf5_d <span class="ot">&lt;-</span> test<span class="sc">$</span>SVD<span class="sc">$</span>data<span class="sc">$</span>d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a><span class="co"># get data</span></span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>omdata <span class="ot">&lt;-</span> test<span class="sc">$</span>OMICS<span class="sc">$</span>data</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Results in hdf5 file for d</span></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a>svd_hdf5_d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">131.8042</span> <span class="fl">131.1643</span> <span class="fl">130.6787</span> <span class="fl">130.6179</span> <span class="fl">130.5750</span> <span class="fl">130.3000</span> <span class="fl">130.1092</span></span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>svd <span class="ot">&lt;-</span> <span class="fu">svd</span>(<span class="fu">scale</span>(omdata))</span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a>svd<span class="sc">$</span>d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">131.8042</span> <span class="fl">131.1643</span> <span class="fl">130.6787</span> <span class="fl">130.6179</span> <span class="fl">130.5750</span> <span class="fl">130.3000</span> <span class="fl">130.1092</span></span></code></pre></div>
<p>Like in Simple Singular Value Decomposition we can normalize, center
or scale data before proceed with SVD decomposition with
<code>bscale</code> and <code>bcenter</code> parameters, by default this
parameter are TRUE, data is normalized before SVD decomposition. To
proceed with SVD without normalization :</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Direct from hdf5 data file (using only one thread, serial execution)</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>svdh5 <span class="ot">&lt;-</span> <span class="fu">bdSVD_hdf5</span>( <span class="st">"delayed.hdf5"</span>, <span class="st">"OMICS"</span>, <span class="st">"data"</span>, </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">bcenter =</span> <span class="cn">FALSE</span>, <span class="at">bscale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">k =</span> <span class="dv">2</span>, <span class="at">q =</span> <span class="dv">1</span>, <span class="at">threads =</span> <span class="dv">1</span>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>SVD <span class="sc">-</span> old dataset have been REMOVED </span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get results svd (d)</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>svd_hdf5_d <span class="ot">&lt;-</span> test<span class="sc">$</span>SVD<span class="sc">$</span>data<span class="sc">$</span>d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>] </span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co"># SVD (d) from file - data not normalized</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>svd_hdf5_d</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">132.0820</span> <span class="fl">131.4080</span> <span class="fl">130.9686</span> <span class="fl">130.8555</span> <span class="fl">130.6781</span> <span class="fl">130.4269</span> <span class="fl">130.3248</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="co"># with R implementation from data in memory</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>svd <span class="ot">&lt;-</span> <span class="fu">svd</span>(omdata)</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>svd<span class="sc">$</span>d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">132.0820</span> <span class="fl">131.4080</span> <span class="fl">130.9686</span> <span class="fl">130.8555</span> <span class="fl">130.6781</span> <span class="fl">130.4269</span> <span class="fl">130.3248</span></span></code></pre></div>
<p>In the SVD decomposition by blocks we can indicate the number of
decomposition levels and number of local SVDs to concatenate at each
level with parameters <code>q</code> and <code>k</code> respectively, by
default <code>q = 1</code> one level with <code>k=2</code>.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Block decomposition with 1 level and 4 local SVDs at each level</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>svdh5 <span class="ot">&lt;-</span> <span class="fu">bdSVD_hdf5</span>(<span class="at">file =</span> <span class="st">"delayed.hdf5"</span>, <span class="at">group =</span> <span class="st">"OMICS"</span>, <span class="at">dataset =</span> <span class="st">"data"</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">threads =</span> <span class="dv">2</span>)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>SVD <span class="sc">-</span> old dataset have been REMOVED </span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="co"># get results svd (d)</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>fprova <span class="ot">&lt;-</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>  svd_hdf5_d <span class="ot">&lt;-</span> fprova<span class="sc">$</span>SVD<span class="sc">$</span>data<span class="sc">$</span>d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="co"># SVD (d) from file - data not normalized</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>svd_hdf5_d</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">131.8042</span> <span class="fl">131.1643</span> <span class="fl">130.6787</span> <span class="fl">130.6179</span> <span class="fl">130.5750</span> <span class="fl">130.3000</span> <span class="fl">130.1092</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="co"># with R implementation from data in memory</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>svd <span class="ot">&lt;-</span> <span class="fu">svd</span>(<span class="fu">scale</span>(omdata))</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>svd<span class="sc">$</span>d[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">131.8042</span> <span class="fl">131.1643</span> <span class="fl">130.6787</span> <span class="fl">130.6179</span> <span class="fl">130.5750</span> <span class="fl">130.3000</span> <span class="fl">130.1092</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="utils-for-data-analysis-working-directly-in-hdf5-files-in-genomic-data">Utils for data analysis working directly in hdf5 files in genomic
data<a class="anchor" aria-label="anchor" href="#utils-for-data-analysis-working-directly-in-hdf5-files-in-genomic-data"></a>
</h2>
<p>We have also implemented a set of functions to deal with genomic data
which is a setting were big data is normally present. These are
described in the next sections</p>
<div class="section level3">
<h3 id="genomic-data-imputation">Genomic data imputation<a class="anchor" aria-label="anchor" href="#genomic-data-imputation"></a>
</h3>
<p>Imputation in genetics refers to the statistical inference of
unobserved genotypes. In genetic epidemiology and quantitative genetics,
researchers aim at identifying genomic locations where variation between
individuals is associated with variation in traits of interest between
individuals.</p>
<p>At the moment <code>BigDataStatMeth</code> has implemented snps
imputation following the encoding used in Genomic Data Structure file
format (GDS files) where SNPs are codded as 0 : two B alleles, 1 : one A
allele and one B allele, 2 : two A alleles and 3 : missing data.
BigDataStatMeth impute data where SNPs values are ‘3’ (missing
data).</p>
<p>The imputation method implemented is generated from the information
of the SNP data that contains the missing data, it is performed by
generating a random value following a discrete distribution. If in the
SNP we find 70% of ‘0’, 25% of ‘1’ and 5% of ‘2’, the value of the
missing data will be imputed with a probability of 70% to be ‘0’, 25% to
be ‘1’ and 5% to be ‘2’.</p>
<p>We first simulate a genotype matrix with 0, 1, 2 and 3 and store it
in hdf5 file and show data stored in file:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">108432</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>geno.sim <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">sample</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">10000</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>), <span class="at">byrow =</span> <span class="cn">TRUE</span>, <span class="at">ncol =</span> <span class="dv">20</span>)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">bdAdd_hdf5_matrix</span>(geno.sim, <span class="st">"delayed.hdf5"</span>, <span class="st">"OMICS"</span>, <span class="st">"geno"</span>)</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>Dataset has been added to the file</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get data and show the first 5 rows</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>h5fsvd <span class="ot">=</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>geno <span class="ot">&lt;-</span> h5fsvd<span class="sc">$</span>OMICS<span class="sc">$</span>geno</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>geno[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">1</span>] [,<span class="dv">2</span>] [,<span class="dv">3</span>] [,<span class="dv">4</span>] [,<span class="dv">5</span>] [,<span class="dv">6</span>] [,<span class="dv">7</span>] [,<span class="dv">8</span>] [,<span class="dv">9</span>] [,<span class="dv">10</span>]</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">1</span>     <span class="dv">0</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]    <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">3</span>    <span class="dv">1</span>     <span class="dv">2</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,]    <span class="dv">3</span>    <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">3</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">3</span>    <span class="dv">1</span>     <span class="dv">1</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]    <span class="dv">3</span>    <span class="dv">3</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">0</span>     <span class="dv">3</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">3</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">2</span>     <span class="dv">1</span></span></code></pre></div>
<p>Remember, we allways have to close hdf5 data file. Now we impute
values where genotype = ‘3’ (missing data in GDS format). To apply
imputation we only have to indicate the file where dataset is, the group
and the dataset name, optionally, we indicate were we want to store
results with <code>outputgroup</code> and <code>outputfolder</code>, by
default the results are stored in the input dataset. An important
parameter is <code>bycols</code>, with <code>bycols</code> we can inform
if we want to impute data by cols <code>bycols = TRUE</code> or by rows
<code>bycols = FALSE</code>, the default imputation implemented method
is by cols.</p>
<p>In next example we perform the imputation in “imputedgeno” dataset
inside the group “OMICS” and the imputation method applied takes into
account the probabilities in columns (default).</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bdImpute_snps_hdf5</span>(<span class="st">"delayed.hdf5"</span>, <span class="at">group=</span><span class="st">"OMICS"</span>, <span class="at">dataset=</span><span class="st">"geno"</span>,</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">outgroup=</span><span class="st">"OMICS"</span>, <span class="at">outdataset=</span><span class="st">"imputedgeno"</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>SNPs with missing values has been imputed</span></code></pre></div>
<p>Now we get the imputation results</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get imputed data and show the first 5 rows</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>h5fsvd <span class="ot">=</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>imputedgeno <span class="ot">&lt;-</span> h5fsvd<span class="sc">$</span>OMICS<span class="sc">$</span>imputedgeno</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>imputedgeno[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">1</span>] [,<span class="dv">2</span>] [,<span class="dv">3</span>] [,<span class="dv">4</span>] [,<span class="dv">5</span>] [,<span class="dv">6</span>] [,<span class="dv">7</span>] [,<span class="dv">8</span>] [,<span class="dv">9</span>] [,<span class="dv">10</span>]</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">1</span>     <span class="dv">0</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]    <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">1</span>     <span class="dv">2</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,]    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">1</span>     <span class="dv">1</span></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">0</span>     <span class="dv">2</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">2</span>     <span class="dv">1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="normalization-center-and-scale-of-genomic-data">Normalization (center and scale) of genomic data<a class="anchor" aria-label="anchor" href="#normalization-center-and-scale-of-genomic-data"></a>
</h3>
<p>Normalization is one of the most important procedures in genomics
data analysis. A typical dataset contains more than one sample and we
are almost always interested in making comparisons between these and in
order to make comparisons we need to normalize data.</p>
<p>In BigDataStatMeth we implemented a normalization method that works
directly with datasets stored in hdf5 file format, like in other
functions implemented in BigDataSet that works with hdf5 data files, we
have to indicate where the data is, the filename, the name of the group
under the dataset is and the dataset name. In normalization function, we
can indicate if we want to center data, scale or center and scale
(default option). The applied formula to normalize data by default in
implemented function is : <span class="math display">\[ \frac{X -
\mu}{\sigma}\]</span></p>
<p>The implemented function to normalize data direct from hdf5 data file
is <code>bdNormalize_hdf5</code>.</p>
<p>To show you an example, we will use the imputed geno dataset
<em>imputedgeno</em> created before in imputation method example. To
normalize data we have to indicate where the dataset is stored and we
have two optiona paramters <code>bcenter</code> and <code>bscale</code>
with default value = TRUE. In order to normalize data, we don’t need to
modify these optional parameters. If we only want to center data, then
we have to put <code>bscale = FALSE</code> or if we only want to scale
data <code>bcenter</code> parameter must be FALSE.</p>
<p><strong>Important : </strong> The normalization results are stored
under “NORMALIZED” group, under this group, we found all normalized
datasets with the same structure under root, for example, if we
normalize dataset genoimputed under OMICS group, the normalized results
will be in “NORMALIZED”/“OMICS”/genoimputed</p>
<p>Let show an example normalizing imputed geno under group “OMICS” in
delayed.hdf5 file:</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">bdNormalize_hdf5</span>(<span class="st">"delayed.hdf5"</span>, <span class="at">group=</span><span class="st">"OMICS"</span>, <span class="at">dataset=</span><span class="st">"imputedgeno"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>Normalization has been computed</span></code></pre></div>
<p>The results will be under groups “NORMALIZED” - “OMICS” in dataset
genoimputed</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Geno after normalization</span></span>
<span><span class="va">h5fsvd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/H5Fopen.html" class="external-link">H5Fopen</a></span><span class="op">(</span><span class="st">"delayed.hdf5"</span><span class="op">)</span></span>
<span><span class="va">genonormalized</span> <span class="op">&lt;-</span> <span class="va">h5fsvd</span><span class="op">$</span><span class="va">NORMALIZED</span><span class="op">$</span><span class="va">OMICS</span><span class="op">$</span><span class="va">geno</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rhdf5/man/h5closeAll.html" class="external-link">h5closeAll</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genonormalized</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="cn">NULL</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="qc---remove-low-data">QC - Remove low data<a class="anchor" aria-label="anchor" href="#qc---remove-low-data"></a>
</h3>
<p>Missing data in meta-analysis of genetic association studies are
unavoidable. In BigDataStatMeth we implemented a function to remove
those SNPs with a hight percentage of missing data. The implemented
function to remove SNPs wit a certain percentage of missing data is the
function <code>bdRemovelowdata</code></p>
<p>Like in imputation method, in remove low data method we have to
inform the input data indicating the filename, the group and the dataset
name and we also have to inform the dataset were data will be stored
after removing the SNPs with parameters <code>outgroup</code> and
<code>outdataset</code>.</p>
<p>To remove low data we have to infrorm if we want to remove data by
cols <code>bycols = TRUE</code> or by rows <code>bycols = FALSE</code>
and the percentage of missing data <code>pcent</code> parameter.
<code>pcent</code> of missing data refers to missing percentage in
columns if we defined <code>bycols = TRUE</code> or <code>pcent</code>
refers to rows if <code>bycols = FALSE</code>.</p>
<p>To show how renove low data works we use the previous dataset stored
in <em>delayed.hdf5</em> file, in that case, we will assume that SNPs
are in columns and we will remove those SNPs where missing data is
greater than 40%.</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bdRemovelowdata</span>(<span class="st">"delayed.hdf5"</span>, <span class="at">group=</span><span class="st">"OMICS"</span>, <span class="at">dataset=</span><span class="st">"geno"</span>,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">outgroup=</span><span class="st">"OMICS"</span>, <span class="at">outdataset=</span><span class="st">"removedgeno"</span>,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">bycols =</span> <span class="cn">TRUE</span>, <span class="at">pcent =</span> <span class="fl">0.39</span>)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>Warning <span class="cf">in</span> <span class="fu">evalq</span>((<span class="cf">function</span> (..., <span class="at">call. =</span> <span class="cn">TRUE</span>, <span class="at">immediate. =</span> <span class="cn">FALSE</span>, <span class="at">noBreaks. =</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="cn">FALSE</span>, <span class="sc">:</span> <span class="dv">0</span> Columns have been removed</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">0</span></span></code></pre></div>
<p>After remove the SNPs with pcent greater than 39% results have been
saved in dataset “<em>removedgeno</em>” under group “<em>OMICS</em>”, we
observe the resulting dataset</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get imputed data and show the first 5 rows</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>h5fsvd <span class="ot">=</span> <span class="fu">H5Fopen</span>(<span class="st">"delayed.hdf5"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>removedgeno <span class="ot">&lt;-</span> h5fsvd<span class="sc">$</span>OMICS<span class="sc">$</span>removedgeno</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="fu">h5closeAll</span>()</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>removedgeno[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>     [,<span class="dv">1</span>] [,<span class="dv">2</span>] [,<span class="dv">3</span>] [,<span class="dv">4</span>] [,<span class="dv">5</span>] [,<span class="dv">6</span>] [,<span class="dv">7</span>] [,<span class="dv">8</span>] [,<span class="dv">9</span>] [,<span class="dv">10</span>]</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">1</span>     <span class="dv">0</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,]    <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">3</span>    <span class="dv">1</span>     <span class="dv">2</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>[<span class="dv">3</span>,]    <span class="dv">3</span>    <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">3</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">3</span>    <span class="dv">1</span>     <span class="dv">1</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>,]    <span class="dv">3</span>    <span class="dv">3</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">2</span>    <span class="dv">1</span>    <span class="dv">0</span>     <span class="dv">3</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>[<span class="dv">5</span>,]    <span class="dv">0</span>    <span class="dv">2</span>    <span class="dv">0</span>    <span class="dv">3</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">2</span>     <span class="dv">1</span></span></code></pre></div>
<!-- ## QC - Get MAF -->
<!-- ## QC - Remove MAF -->
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>R version <span class="dv">4</span>.<span class="fl">2.1</span> (<span class="dv">2022-06-23</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>Platform<span class="sc">:</span> aarch64<span class="sc">-</span>apple<span class="sc">-</span><span class="fu">darwin20</span> (<span class="dv">64</span><span class="sc">-</span>bit)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>Running under<span class="sc">:</span> macOS Monterey <span class="fl">12.5</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>Matrix products<span class="sc">:</span> default</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>BLAS<span class="sc">:</span>   <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.2</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRblas.<span class="fl">0.</span>dylib</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>LAPACK<span class="sc">:</span> <span class="er">/</span>Library<span class="sc">/</span>Frameworks<span class="sc">/</span>R.framework<span class="sc">/</span>Versions<span class="sc">/</span><span class="fl">4.2</span><span class="sc">-</span>arm64<span class="sc">/</span>Resources<span class="sc">/</span>lib<span class="sc">/</span>libRlapack.dylib</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>locale<span class="sc">:</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] ca_ES.UTF<span class="dv">-8</span><span class="sc">/</span>ca_ES.UTF<span class="dv">-8</span><span class="sc">/</span>ca_ES.UTF<span class="dv">-8</span><span class="sc">/</span>C<span class="sc">/</span>ca_ES.UTF<span class="dv">-8</span><span class="sc">/</span>ca_ES.UTF<span class="dv">-8</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>attached base packages<span class="sc">:</span></span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>other attached packages<span class="sc">:</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] BigDataStatMeth_0.<span class="fl">99.50</span> rhdf5_2.<span class="fl">40.0</span>            knitr_1<span class="fl">.40</span>             </span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>[<span class="dv">4</span>] BiocStyle_2.<span class="fl">24.0</span>       </span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>loaded via a <span class="fu">namespace</span> (and not attached)<span class="sc">:</span></span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a> [<span class="dv">1</span>] Rcpp_1.<span class="fl">0.9</span>          bslib_0.<span class="fl">4.0</span>         compiler_4.<span class="fl">2.1</span>     </span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a> [<span class="dv">4</span>] BiocManager_1.<span class="fl">30.18</span> jquerylib_0.<span class="fl">1.4</span>     highr_0<span class="fl">.9</span>          </span>
<span id="cb41-23"><a href="#cb41-23" aria-hidden="true" tabindex="-1"></a> [<span class="dv">7</span>] rhdf5filters_1.<span class="fl">8.0</span>  bitops_1<span class="fl">.0</span><span class="dv">-7</span>        tools_4.<span class="fl">2.1</span>        </span>
<span id="cb41-24"><a href="#cb41-24" aria-hidden="true" tabindex="-1"></a>[<span class="dv">10</span>] digest_0.<span class="fl">6.29</span>       jsonlite_1.<span class="fl">8.2</span>      evaluate_0<span class="fl">.17</span>      </span>
<span id="cb41-25"><a href="#cb41-25" aria-hidden="true" tabindex="-1"></a>[<span class="dv">13</span>] memoise_2.<span class="fl">0.1</span>       rlang_1.<span class="fl">0.6</span>         cli_3.<span class="fl">4.1</span>          </span>
<span id="cb41-26"><a href="#cb41-26" aria-hidden="true" tabindex="-1"></a>[<span class="dv">16</span>] rstudioapi_0<span class="fl">.14</span>     yaml_2.<span class="fl">3.5</span>          pkgdown_2.<span class="fl">0.6</span>      </span>
<span id="cb41-27"><a href="#cb41-27" aria-hidden="true" tabindex="-1"></a>[<span class="dv">19</span>] xfun_0<span class="fl">.33</span>           fastmap_1.<span class="fl">1.0</span>       stringr_1.<span class="fl">4.1</span>      </span>
<span id="cb41-28"><a href="#cb41-28" aria-hidden="true" tabindex="-1"></a>[<span class="dv">22</span>] desc_1.<span class="fl">4.2</span>          fs_1.<span class="fl">5.2</span>            sass_0.<span class="fl">4.2</span>         </span>
<span id="cb41-29"><a href="#cb41-29" aria-hidden="true" tabindex="-1"></a>[<span class="dv">25</span>] systemfonts_1.<span class="fl">0.4</span>   rprojroot_2.<span class="fl">0.3</span>     data.table_1.<span class="fl">14.2</span>  </span>
<span id="cb41-30"><a href="#cb41-30" aria-hidden="true" tabindex="-1"></a>[<span class="dv">28</span>] R6_2.<span class="fl">5.1</span>            textshaping_0.<span class="fl">3.6</span>   rmarkdown_2<span class="fl">.17</span>     </span>
<span id="cb41-31"><a href="#cb41-31" aria-hidden="true" tabindex="-1"></a>[<span class="dv">31</span>] bookdown_0<span class="fl">.29</span>       purrr_0.<span class="fl">3.5</span>         Rhdf5lib_1.<span class="fl">18.2</span>    </span>
<span id="cb41-32"><a href="#cb41-32" aria-hidden="true" tabindex="-1"></a>[<span class="dv">34</span>] magrittr_2.<span class="fl">0.3</span>      htmltools_0.<span class="fl">5.3</span>     ragg_1.<span class="fl">2.3</span>         </span>
<span id="cb41-33"><a href="#cb41-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">37</span>] stringi_1.<span class="fl">7.8</span>       RCurl_1<span class="fl">.98-1.9</span>      RcppParallel_5.<span class="fl">1.5</span> </span>
<span id="cb41-34"><a href="#cb41-34" aria-hidden="true" tabindex="-1"></a>[<span class="dv">40</span>] cachem_1.<span class="fl">0.6</span>       </span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Dolors Pelegri-Siso, Juan R. Gonzalez.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
